## Load Data from Table

In [0]:
%sql
USE andres_osoriogr_outlook_com_db

In [0]:
retail_df = spark.read.table("retail_table")
retail_df.printSchema()

root
-- invoice_no: string (nullable = true)
-- stock_code: string (nullable = true)
-- description: string (nullable = true)
-- quantity: integer (nullable = true)
-- invoice_date: timestamp (nullable = true)
-- unit_price: double (nullable = true)
-- customer_id: integer (nullable = true)
-- country: string (nullable = true)

In [0]:
print("Partitions: " + str(retail_df.rdd.getNumPartitions()))

Partitions: 4

In [0]:
display(retail_df.select("*").limit(10))

invoice_no stock_code description quantity invoice_date unit_price customer_id country 489434 85048 15CM CHRISTMAS GLASS BALL 20 LIGHTS 12 2009-12-01T07:45:00.000+0000 6.95 13085 United Kingdom 489434 79323P PINK CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 79323W WHITE CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 22041 RECORD FRAME 7" SINGLE SIZE 48 2009-12-01T07:45:00.000+0000 2.1 13085 United Kingdom 489434 21232 STRAWBERRY CERAMIC TRINKET BOX 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 22064 PINK DOUGHNUT TRINKET POT 24 2009-12-01T07:45:00.000+0000 1.65 13085 United Kingdom 489434 21871 SAVE THE PLANET MUG 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 21523 FANCY FONT HOME SWEET HOME DOORMAT 10 2009-12-01T07:45:00.000+0000 5.95 13085 United Kingdom 489435 22350 CAT BOWL 12 2009-12-01T07:46:00.000+0000 2.55 13085 United Kingdom 489435 22349 DOG BOWL , CHASING BALL DESIGN 12 2009-12-01T07:46:00.000+0000 3.75 13085 United Kingdom

In [0]:
%sql
SELECT * FROM retail_table LIMIT 10

invoice_no stock_code description quantity invoice_date unit_price customer_id country 489434 85048 15CM CHRISTMAS GLASS BALL 20 LIGHTS 12 2009-12-01T07:45:00.000+0000 6.95 13085 United Kingdom 489434 79323P PINK CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 79323W WHITE CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 22041 RECORD FRAME 7" SINGLE SIZE 48 2009-12-01T07:45:00.000+0000 2.1 13085 United Kingdom 489434 21232 STRAWBERRY CERAMIC TRINKET BOX 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 22064 PINK DOUGHNUT TRINKET POT 24 2009-12-01T07:45:00.000+0000 1.65 13085 United Kingdom 489434 21871 SAVE THE PLANET MUG 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 21523 FANCY FONT HOME SWEET HOME DOORMAT 10 2009-12-01T07:45:00.000+0000 5.95 13085 United Kingdom 489435 22350 CAT BOWL 12 2009-12-01T07:46:00.000+0000 2.55 13085 United Kingdom 489435 22349 DOG BOWL , CHASING BALL DESIGN 12 2009-12-01T07:46:00.000+0000 3.75 13085 United Kingdom

## Total Invoice Amount Distribution
- Each order (invoice_no) has multiple items
- Calculate the sum of (quantity * unit_price) for each order
- Some amounts are too high and skew the distribution; therefore, remove outliers

#### Get the amount for each row, get only relevant columns, and remove rows with negative amounts

In [0]:
from pyspark.sql.functions import *

invoice_amount_df = retail_df.withColumn("amount", expr("quantity * unit_price"))
invoice_amount_df = invoice_amount_df.select("invoice_no", "amount")
invoice_amount_df = invoice_amount_df.where("amount > 0")

#### Sum up all amounts for each invoice_no

In [0]:
total_amounts_df = invoice_amount_df.groupBy("invoice_no").sum("amount")
total_amounts_df = total_amounts_df.withColumnRenamed("sum(amount)", "total_amount")

#### Distribution with outliers (a few very big amounts)

In [0]:
display(total_amounts_df.describe("total_amount"))

summary total_amount count 40078 mean 523.3037611158244 stddev 1517.351645669805 min 0.19 max 168469.6

In [0]:
display(total_amounts_df.orderBy("invoice_no"))

invoice_no total_amount 489434 505.30000000000007 489435 145.79999999999998 489436 630.33 489437 310.75 489438 2286.24 489439 426.29999999999995 489440 50.4 489441 344.34000000000003 489442 382.37000000000006 489443 285.06 489444 141.0 489445 308.44 489446 996.1 489447 130.0 489448 570.24 489450 196.10000000000002 489460 485.7900000000001 489461 589.33 489462 148.3 489465 1197.8 489488 251.1 489505 255.49999999999997 489514 823.02 489517 282.46000000000004 489519 696.7300000000002 489520 733.6300000000002 489522 700.3099999999998 489523 1160.0 489525 1.4 489526 929.9200000000001 489529 229.45999999999998 489531 144.99999999999997 489532 304.53 489533 192.74999999999997 489536 367.41 489537 633.3300000000003 489539 2887.14 489540 537.96 489542 387.6 489544 281.82 489545 124.95 489546 584.5000000000001 489547 641.4499999999999 489548 390.86 489550 482.2999999999999 489552 278.6099999999999 489556 379.64999999999986 489557 603.94 489558 291.58000000000004 489559 128.60000000000002 489560 859.1 489561 246.86 489562 229.6 489564 322.8999999999999 489566 218.54999999999998 489567 307.79999999999995 489569 282.28 489571 156.70999999999998 489572 507.48999999999967 489573 580.02 489574 551.3299999999998 489575 367.2799999999999 489576 319.66 489579 219.81999999999996 489580 261.0 489581 124.83000000000001 489582 293.95 489584 21.95 489585 105.85 489586 263.42 489590 362.78 489593 423.86999999999995 489594 322.33000000000004 489596 17.0 489597 5097.82 489598 822.37 489599 2454.68 489600 420.29000000000013 489601 337.2 489602 440.07 489603 311.55 489604 383.87000000000006 489606 2.1 489607 446.98000000000013 489608 9.75 489609 553.7800000000001 489611 2.95 489612 629.7800000000002 489613 12.75 489614 408.85 489615 499.36 489616 409.27000000000004 489617 198.0 489618 6.35 489619 19.5 489634 17.4 489635 352.85 489637 535.9 489638 7.5 489639 31.200000000000003 489640 329.90999999999997 489641 622.1 489642 12.850000000000001 489644 896.54 489645 261.74 489650 173.10000000000002 489652 294.14 489653 60.78 489654 222.12000000000003 489656 1237.3 489657 55.2 489658 828.0899999999999 489661 23.48 489662 73.75 489664 374.43 489665 252.3 489666 263.09000000000003 489667 538.46 489668 378.11999999999995 489669 397.20000000000005 489670 114.60000000000001 489671 740.3 489672 772.6499999999999 489673 61.099999999999994 489674 261.3 489675 4939.219999999999 489676 1210.8 489677 192.0 489678 189.29999999999998 489679 795.9800000000001 489680 45.0 489681 89.58999999999999 489682 125.72999999999999 489683 1317.9700000000005 489684 266.6 489687 221.53 489688 299.70000000000005 489695 306.95 489699 401.7 489702 3108.640000000001 489705 17.0 489723 2052.12 489725 216.0 489726 750.0 489727 900.0 489728 180.45 489733 168.0 489743 538.7699999999999 489748 754.2 489764 223.09999999999997 489765 285.5900000000001 489766 130.17999999999998 489780 159.98 489782 230.46 489783 162.04000000000002 489784 40.8 489785 333.67 489786 216.86 489787 319.5399999999999 489788 119.16 489789 80.4 489790 131.12 489791 221.18999999999988 489792 923.7900000000003 489793 260.36 489794 58.2 489795 935.0100000000002 489796 311.5 489797 200.63 489798 311.69 489800 408.06999999999994 489801 128.48000000000002 489802 174.52000000000004 489804 251.17000000000002 489805 454.35 489807 180.12999999999997 489808 109.84 489809 112.5 489810 493.2500000000001 489811 634.29 489812 597.6899999999999 489814 824.1800000000004 489815 132.55 489816 284.54999999999995 489817 340.2 489818 294.3 489819 618.43 489822 371.27000000000004 489823 87.30000000000001 489824 176.87999999999997 489825 174.1 489826 85.95 489827 445.7000000000001 489829 472.46999999999997 489830 637.0399999999998 489831 1008.0 489832 316.21999999999997 489833 150.68 489834 214.79999999999998 489835 714.0 489837 514.97 489840 89.69999999999999 489855 157.90999999999997 489856 4463.740000000002 489857 5741.960000000003 489861 46.07000000000001 489862 734.7 489863 66.82000000000001 489864 206.85 489865 234.87 489866 1003.829999999999

#### Distribution without outliers (only consider the first ~60% quantile of original data)

In [0]:
quantile = total_amounts_df.approxQuantile("total_amount", [0.6], 0.1)
cut = quantile[0]
print(cut)

436.04

In [0]:
no_outliers_df = total_amounts_df.filter(col("total_amount") < cut)
display(no_outliers_df.describe("total_amount"))

summary total_amount count 27980 mean 209.93644270907768 stddev 120.53792537422213 min 0.19 max 436.02999999999986

In [0]:
display(no_outliers_df.orderBy("invoice_no"))

invoice_no total_amount 489435 145.79999999999998 489437 310.75 489439 426.29999999999995 489440 50.4 489441 344.34000000000003 489442 382.37000000000006 489443 285.06 489444 141.0 489445 308.44 489447 130.0 489450 196.10000000000002 489462 148.3 489488 251.1 489505 255.49999999999997 489517 282.46000000000004 489525 1.4 489529 229.45999999999998 489531 144.99999999999997 489532 304.53 489533 192.74999999999997 489536 367.41 489542 387.6 489544 281.82 489545 124.95 489548 390.86 489552 278.6099999999999 489556 379.64999999999986 489558 291.58000000000004 489559 128.60000000000002 489561 246.86 489562 229.6 489564 322.8999999999999 489566 218.54999999999998 489567 307.79999999999995 489569 282.28 489571 156.70999999999998 489575 367.2799999999999 489576 319.66 489579 219.81999999999996 489580 261.0 489581 124.83000000000001 489582 293.95 489584 21.95 489585 105.85 489586 263.42 489590 362.78 489593 423.86999999999995 489594 322.33000000000004 489596 17.0 489600 420.29000000000013 489601 337.2 489603 311.55 489604 383.87000000000006 489606 2.1 489608 9.75 489611 2.95 489613 12.75 489614 408.85 489616 409.27000000000004 489617 198.0 489618 6.35 489619 19.5 489634 17.4 489635 352.85 489638 7.5 489639 31.200000000000003 489640 329.90999999999997 489642 12.850000000000001 489645 261.74 489650 173.10000000000002 489652 294.14 489653 60.78 489654 222.12000000000003 489657 55.2 489661 23.48 489662 73.75 489664 374.43 489665 252.3 489666 263.09000000000003 489668 378.11999999999995 489669 397.20000000000005 489670 114.60000000000001 489673 61.099999999999994 489674 261.3 489677 192.0 489678 189.29999999999998 489680 45.0 489681 89.58999999999999 489682 125.72999999999999 489684 266.6 489687 221.53 489688 299.70000000000005 489695 306.95 489699 401.7 489705 17.0 489725 216.0 489728 180.45 489733 168.0 489764 223.09999999999997 489765 285.5900000000001 489766 130.17999999999998 489780 159.98 489782 230.46 489783 162.04000000000002 489784 40.8 489785 333.67 489786 216.86 489787 319.5399999999999 489788 119.16 489789 80.4 489790 131.12 489791 221.18999999999988 489793 260.36 489794 58.2 489796 311.5 489797 200.63 489798 311.69 489800 408.06999999999994 489801 128.48000000000002 489802 174.52000000000004 489804 251.17000000000002 489807 180.12999999999997 489808 109.84 489809 112.5 489815 132.55 489816 284.54999999999995 489817 340.2 489818 294.3 489822 371.27000000000004 489823 87.30000000000001 489824 176.87999999999997 489825 174.1 489826 85.95 489832 316.21999999999997 489833 150.68 489834 214.79999999999998 489840 89.69999999999999 489855 157.90999999999997 489861 46.07000000000001 489863 66.82000000000001 489864 206.85 489865 234.87 489867 118.95 489869 39.98 489870 162.1 489873 10.2 489874 176.6 489876 39.85000000000001 489877 368.1 489878 143.11000000000004 489879 163.46999999999997 489880 380.27999999999986 489884 3.61 489891 181.08 489892 39.8 489893 300.1 489895 371.82 489896 337.93000000000006 489897 260.44 489900 287.37 489902 119.10000000000001 489904 203.77999999999997 489906 120.0 489907 202.06 489908 107.33 489910 144.0 489913 69.6 489915 54.599999999999994 489916 336.0 489932 82.60000000000001 489935 181.65 489994 430.9 489996 320.40000000000003 489998 321.65000000000003 489999 154.20000000000002 490000 143.06 490001 169.5 490002 114.55000000000001 490003 317.54999999999995 490006 344.4 490009 285.09000000000003 490011 221.69000000000003 490017 393.6499999999999 490019 308.6 490020 37.8 490021 10.2 490022 45.60000000000001 490023 189.15 490024 214.99999999999997 490030 58.2 490034 69.82 490045 180.39000000000001 490046 228.3 490047 263.84999999999997 490049 102.0 490050 13.5 490053 119.54 490054 277.5 490056 61.120000000000005 490057 360.5200000000001 490058 47.85 490060 314.21999999999997 490062 165.3 490063 210.38999999999996 490064 355.59000000000003 490066 372.2 490068 284.3 490069 161.39999999999998 490070 236.25 490071 294.85 490072 194.64999999999998 490073 258.09999999999997 490075 381.0 490077 115.50999999999999 49

## Monthly Placed and Cancelled Orders
- The column "invoice_no" denotes the order numbers
- Each order has multiple items (we only care about one appearance of each for this query)
- Orders that start with a 'C' have been cancelled
- Calculate the number of placed VS cancelled orders for each month

#### Create YYYYMM column, get relevant columns only, and remove duplicate invoice_no

In [0]:
orders_df = retail_df.select("*", (year("invoice_date") * 100 + month("invoice_date")).alias("YYYYMM"))
orders_df = orders_df.select("YYYYMM", "invoice_no")
orders_df = orders_df.dropDuplicates(["invoice_no"])
display(orders_df)

YYYYMM invoice_no 200912 489677 200912 C491017 200912 491045 200912 491658 200912 C491705 200912 C492541 200912 C493168 201001 493542 201001 493977 201001 C493984 201001 494244 201001 494277 201001 C494502 201001 495185 201001 495329 201001 495594 201001 C495697 201001 495783 201001 496171 201001 496233 201002 496427 201002 497229 201002 C497585 201002 497815 201002 498059 201002 498070 201002 498125 201002 498328 201002 498934 201002 499145 201002 C499380 201003 C499978 201003 500148 201003 500253 201003 500903 201003 500979 201003 501046 201003 501287 201003 501527 201003 501953 201003 C502084 201003 502610 201003 502620 201003 502860 201004 C503733 201004 503942 201004 504372 201004 505024 201004 505692 201004 505703 201004 505751 201004 506092 201004 506258 201005 C506881 201005 507206 201005 507464 201005 508020 201005 508195 201005 508528 201005 508647 201005 508840 201005 508905 201005 508953 201005 509792 201005 C509853 201005 509891 201005 510434 201005 510436 201006 510516 201006 510761 201006 511135 201006 511472 201006 511668 201006 C511707 201006 511761 201006 511789 201006 513264 201006 513499 201006 513575 201006 513652 201006 514019 201006 C514145 201006 C514149 201007 514838 201007 515016 201007 515937 201003 501304 201004 504328 201005 507425 201006 513470 201007 514951 201007 517114 201007 517254 201007 C517375 201008 517742 201008 C517763 201008 518382 201008 518388 201008 518493 201008 518656 201008 518753 201008 519251 201008 519471 201008 519715 201008 C520060 201008 520538 201008 520686 201008 520705 201009 C522717 201009 523563 201009 523989 201009 524274 201009 524305 201009 C524395 201010 525091 201010 C525917 201010 526084 201010 526186 201010 526471 201010 527334 201010 527376 201010 527381 201010 527490 201010 527766 201010 C528480 201010 529593 201011 530316 201011 530320 201011 530405 201011 530928 201011 531036 201011 531404 201011 531419 201011 C531807 201011 C531860 201011 C531874 201011 533759 201011 533786 201011 C533814 201011 533895 201011 533990 201011 C534792 201011 C534825 201012 536938 201012 537252 201012 537691 201012 538041 201012 538184 201012 538517 201012 538879 201009 524183 201010 525494 201011 533588 201012 536596 201012 539275 201012 539630 201101 540499 201101 540540 201101 C540850 201101 540976 201101 541432 201101 541518 201101 541783 201101 542026 201101 542375 201101 C542604 201102 543641 201102 544303 201103 545583 201103 C546174 201103 547122 201103 547557 201103 548542 201104 548998 201104 549160 201104 C549913 201104 550469 201104 550617 201104 C550672 201104 550831 201105 551990 201105 552172 201105 552191 201105 552215 201105 552238 201105 552277 201105 552677 201105 552852 201105 553390 201105 553470 201105 553559 201105 553658 201105 553708 201105 553738 201105 554072 201105 554284 201105 554441 201106 C556414 201106 556441 201106 556462 201106 556732 201106 556929 201106 557271 201106 557757 201106 558100 201107 558700 201107 559062 201107 559174 201107 559295 201107 560095 201107 560145 201107 560536 201107 561057 201107 561352 201107 561519 201108 562204 201108 562271 201108 562463 201108 562600 201108 562890 201108 562945 201108 C562959 201108 563020 201105 551692 201105 553835 201105 554630 201105 554792 201109 565747 201109 566248 201109 566431 201109 567163 201109 567695 201109 567879 201109 568222 201109 568711 201109 569020 201110 569560 201110 569823 201110 570234 201110 570264 201110 570281 201110 570592 201110 571010 201110 571906 201110 572049 201110 572458 201110 573020 201110 573256 201110 573409 201110 573726 201111 574592 201111 574844 201111 574966 201111 575091 201111 575671 201111 575948 201111 575961 201111 576059 201111 576112 201111 C576393 201111 577022 201111 C577362 201111 577511 201111 577541 201111 578057 201111 C578132 201111 578292 201111 578459 201111 578941 201112 580490 201112 580739 201112 580906 201112 581138 200912 489460 200912 489812 200912 489824 200912 489915 200912 490100 200912 C490462 200912 490465 200912 490526 200912 

#### Add column with cancelled orders (orders starting with 'C')

In [0]:
placed_cancelled_df = orders_df.withColumn("cancelled", expr("invoice_no LIKE 'C%'"))
display(placed_cancelled_df)

YYYYMM invoice_no cancelled 200912 489677 false 200912 C491017 true 200912 491045 false 200912 491658 false 200912 C491705 true 200912 C492541 true 200912 C493168 true 201001 493542 false 201001 493977 false 201001 C493984 true 201001 494244 false 201001 494277 false 201001 C494502 true 201001 495185 false 201001 495329 false 201001 495594 false 201001 C495697 true 201001 495783 false 201001 496171 false 201001 496233 false 201002 496427 false 201002 497229 false 201002 C497585 true 201002 497815 false 201002 498059 false 201002 498070 false 201002 498125 false 201002 498328 false 201002 498934 false 201002 499145 false 201002 C499380 true 201003 C499978 true 201003 500148 false 201003 500253 false 201003 500903 false 201003 500979 false 201003 501046 false 201003 501287 false 201003 501527 false 201003 501953 false 201003 C502084 true 201003 502610 false 201003 502620 false 201003 502860 false 201004 C503733 true 201004 503942 false 201004 504372 false 201004 505024 false 201004 505692 false 201004 505703 false 201004 505751 false 201004 506092 false 201004 506258 false 201005 C506881 true 201005 507206 false 201005 507464 false 201005 508020 false 201005 508195 false 201005 508528 false 201005 508647 false 201005 508840 false 201005 508905 false 201005 508953 false 201005 509792 false 201005 C509853 true 201005 509891 false 201005 510434 false 201005 510436 false 201006 510516 false 201006 510761 false 201006 511135 false 201006 511472 false 201006 511668 false 201006 C511707 true 201006 511761 false 201006 511789 false 201006 513264 false 201006 513499 false 201006 513575 false 201006 513652 false 201006 514019 false 201006 C514145 true 201006 C514149 true 201007 514838 false 201007 515016 false 201007 515937 false 201003 501304 false 201004 504328 false 201005 507425 false 201006 513470 false 201007 514951 false 201007 517114 false 201007 517254 false 201007 C517375 true 201008 517742 false 201008 C517763 true 201008 518382 false 201008 518388 false 201008 518493 false 201008 518656 false 201008 518753 false 201008 519251 false 201008 519471 false 201008 519715 false 201008 C520060 true 201008 520538 false 201008 520686 false 201008 520705 false 201009 C522717 true 201009 523563 false 201009 523989 false 201009 524274 false 201009 524305 false 201009 C524395 true 201010 525091 false 201010 C525917 true 201010 526084 false 201010 526186 false 201010 526471 false 201010 527334 false 201010 527376 false 201010 527381 false 201010 527490 false 201010 527766 false 201010 C528480 true 201010 529593 false 201011 530316 false 201011 530320 false 201011 530405 false 201011 530928 false 201011 531036 false 201011 531404 false 201011 531419 false 201011 C531807 true 201011 C531860 true 201011 C531874 true 201011 533759 false 201011 533786 false 201011 C533814 true 201011 533895 false 201011 533990 false 201011 C534792 true 201011 C534825 true 201012 536938 false 201012 537252 false 201012 537691 false 201012 538041 false 201012 538184 false 201012 538517 false 201012 538879 false 201009 524183 false 201010 525494 false 201011 533588 false 201012 536596 false 201012 539275 false 201012 539630 false 201101 540499 false 201101 540540 false 201101 C540850 true 201101 540976 false 201101 541432 false 201101 541518 false 201101 541783 false 201101 542026 false 201101 542375 false 201101 C542604 true 201102 543641 false 201102 544303 false 201103 545583 false 201103 C546174 true 201103 547122 false 201103 547557 false 201103 548542 false 201104 548998 false 201104 549160 false 201104 C549913 true 201104 550469 false 201104 550617 false 201104 C550672 true 201104 550831 false 201105 551990 false 201105 552172 false 201105 552191 false 201105 552215 false 201105 552238 false 201105 552277 false 201105 552677 false 201105 552852 false 201105 553390 false 201105 553470 false 201105 553559 false 201105 553658 false 201105 553708 false 201105 553738 false 201105 554072 false 201105 554284 false 201105 554441 false 201106 C556414 true 201106 55644

#### Count the number of placed orders (cancelled = false) for each month

In [0]:
placed_count_df = placed_cancelled_df.filter("cancelled = false")
placed_count_df = placed_count_df.select("YYYYMM", "cancelled")
placed_count_df = placed_count_df.groupBy("YYYYMM").count()
placed_count_df = placed_count_df.withColumnRenamed("count", "placed")
display(placed_count_df)

YYYYMM placed 201108 1459 201011 3093 201101 1216 201004 1588 201003 1960 201103 1665 201112 869 201012 1699 201001 1333 201005 2011 200912 1929 201009 2004 201006 1859 201106 1683 201107 1657 201109 1994 201008 1604 201002 1729 201010 2489 201105 1848 201007 1673 201111 3021 201110 2275 201104 1504 201102 1174

#### Count the number of cancelled orders (cancelled = true) for each month

In [0]:
cancelled_count_df = placed_cancelled_df.filter("cancelled = true")
cancelled_count_df = cancelled_count_df.select("YYYYMM", "cancelled")
cancelled_count_df = cancelled_count_df.groupBy("YYYYMM").count()
cancelled_count_df = cancelled_count_df.withColumnRenamed("count", "cancelled")
display(cancelled_count_df)

YYYYMM cancelled 201108 278 201011 576 201101 260 201004 304 201003 407 201103 318 201112 146 201012 326 201001 300 201005 407 200912 401 201009 371 201006 357 201106 329 201107 270 201109 333 201008 273 201002 240 201010 476 201105 314 201007 344 201111 441 201110 362 201104 240 201102 219

#### Merge placed and cancelled order counts for each month

In [0]:
placed_cancelled_count_df = placed_count_df.join(cancelled_count_df, "YYYYMM", "left_outer")
display(placed_cancelled_count_df.orderBy("YYYYMM"))

YYYYMM placed cancelled 200912 1929 401 201001 1333 300 201002 1729 240 201003 1960 407 201004 1588 304 201005 2011 407 201006 1859 357 201007 1673 344 201008 1604 273 201009 2004 371 201010 2489 476 201011 3093 576 201012 1699 326 201101 1216 260 201102 1174 219 201103 1665 318 201104 1504 240 201105 1848 314 201106 1683 329 201107 1657 270 201108 1459 278 201109 1994 333 201110 2275 362 201111 3021 441 201112 869 146

In [0]:
display(placed_cancelled_count_df.orderBy("YYYYMM"))

YYYYMM placed cancelled 200912 1929 401 201001 1333 300 201002 1729 240 201003 1960 407 201004 1588 304 201005 2011 407 201006 1859 357 201007 1673 344 201008 1604 273 201009 2004 371 201010 2489 476 201011 3093 576 201012 1699 326 201101 1216 260 201102 1174 219 201103 1665 318 201104 1504 240 201105 1848 314 201106 1683 329 201107 1657 270 201108 1459 278 201109 1994 333 201110 2275 362 201111 3021 441 201112 869 146

## Monthly Sales
- A sale is a placed order, therefore we can reuse the dataframe from the last query (placed_count_df)
- Find the maximum number of placed orders and use it to normalize the data on a scale of 0 to 1

In [0]:
display(placed_count_df.describe("placed"))

summary placed count 25 mean 1813.44 stddev 509.0469919368938 min 869 max 3093

In [0]:
sales_df = placed_count_df.select("YYYYMM", (col("placed") / 3093).alias("scaled_sales"))
display(sales_df.orderBy("YYYYMM"))

YYYYMM scaled_sales 200912 0.623666343355965 201001 0.4309731652117685 201002 0.5590042030391206 201003 0.6336889751050759 201004 0.5134173941157453 201005 0.6501778208858713 201006 0.6010345942450696 201007 0.5408988037504041 201008 0.5185903653410928 201009 0.6479146459747818 201010 0.8047203362431297 201011 1.0 201012 0.5493048819915939 201101 0.3931458131264145 201102 0.3795667636598771 201103 0.5383123181377304 201104 0.4862592951826705 201105 0.597478176527643 201106 0.5441319107662463 201107 0.5357258325250566 201108 0.47171031361138055 201109 0.6446815389589395 201110 0.735531846104106 201111 0.976721629485936 201112 0.2809569996766893

In [0]:
display(sales_df.orderBy("YYYYMM"))

YYYYMM scaled_sales 200912 0.623666343355965 201001 0.4309731652117685 201002 0.5590042030391206 201003 0.6336889751050759 201004 0.5134173941157453 201005 0.6501778208858713 201006 0.6010345942450696 201007 0.5408988037504041 201008 0.5185903653410928 201009 0.6479146459747818 201010 0.8047203362431297 201011 1.0 201012 0.5493048819915939 201101 0.3931458131264145 201102 0.3795667636598771 201103 0.5383123181377304 201104 0.4862592951826705 201105 0.597478176527643 201106 0.5441319107662463 201107 0.5357258325250566 201108 0.47171031361138055 201109 0.6446815389589395 201110 0.735531846104106 201111 0.976721629485936 201112 0.2809569996766893

## Monthly Sales Growth
- Calculate the percetange of sales growth for each month
- We can reuse the placed_count_df from before since it has the number of placed orders (sales) for each month
- There's no percetange change function in PySpark; use the lag function to create a new column with the previous value of the number of placed orders and calculate the percetange change this way

In [0]:
from pyspark.sql.window import Window

growth_df = placed_count_df
g_window = Window.partitionBy().orderBy("YYYYMM")
growth_df = growth_df.withColumn("prev_value", lag(growth_df.placed).over(g_window))
growth_df = growth_df.withColumn("growth%", when(isnull(growth_df.placed - growth_df.prev_value), 0)
                              .otherwise((growth_df.placed - growth_df.prev_value) / growth_df.prev_value))
display(growth_df)

YYYYMM placed prev_value growth% 200912 1929 null 0.0 201001 1333 1929 -0.30896837739761535 201002 1729 1333 0.2970742685671418 201003 1960 1729 0.13360323886639677 201004 1588 1960 -0.18979591836734694 201005 2011 1588 0.2663727959697733 201006 1859 2011 -0.07558428642466435 201007 1673 1859 -0.10005379236148466 201008 1604 1673 -0.041243275552898986 201009 2004 1604 0.24937655860349128 201010 2489 2004 0.24201596806387227 201011 3093 2489 0.2426677380474086 201012 1699 3093 -0.45069511800840606 201101 1216 1699 -0.2842848734549735 201102 1174 1216 -0.03453947368421053 201103 1665 1174 0.4182282793867121 201104 1504 1665 -0.0966966966966967 201105 1848 1504 0.22872340425531915 201106 1683 1848 -0.08928571428571429 201107 1657 1683 -0.015448603683897801 201108 1459 1657 -0.11949305974652988 201109 1994 1459 0.3666895133653187 201110 2275 1994 0.14092276830491474 201111 3021 2275 0.3279120879120879 201112 869 3021 -0.712346904998345

In [0]:
display(growth_df)

YYYYMM placed prev_value growth% 200912 1929 null 0.0 201001 1333 1929 -0.30896837739761535 201002 1729 1333 0.2970742685671418 201003 1960 1729 0.13360323886639677 201004 1588 1960 -0.18979591836734694 201005 2011 1588 0.2663727959697733 201006 1859 2011 -0.07558428642466435 201007 1673 1859 -0.10005379236148466 201008 1604 1673 -0.041243275552898986 201009 2004 1604 0.24937655860349128 201010 2489 2004 0.24201596806387227 201011 3093 2489 0.2426677380474086 201012 1699 3093 -0.45069511800840606 201101 1216 1699 -0.2842848734549735 201102 1174 1216 -0.03453947368421053 201103 1665 1174 0.4182282793867121 201104 1504 1665 -0.0966966966966967 201105 1848 1504 0.22872340425531915 201106 1683 1848 -0.08928571428571429 201107 1657 1683 -0.015448603683897801 201108 1459 1657 -0.11949305974652988 201109 1994 1459 0.3666895133653187 201110 2275 1994 0.14092276830491474 201111 3021 2275 0.3279120879120879 201112 869 3021 -0.712346904998345

## Monthly Active Users
- Calculate the number of distinct customer_ids for each month

#### Create 'YYYYMM' column, select relevant columns only, and do the aggregation operation

In [0]:
customers_df = retail_df.select("*", (year("invoice_date") * 100 + month("invoice_date")).alias("YYYYMM"))
customers_df = customers_df.select("YYYYMM", "customer_id")
customers_count_df = customers_df.groupBy("YYYYMM").agg(countDistinct("customer_id").alias("active_users"))
display(customers_count_df.orderBy("YYYYMM"))

YYYYMM active_users 200912 1045 201001 786 201002 807 201003 1111 201004 998 201005 1062 201006 1095 201007 988 201008 964 201009 1202 201010 1577 201011 1683 201012 948 201101 783 201102 798 201103 1020 201104 899 201105 1079 201106 1051 201107 993 201108 980 201109 1302 201110 1425 201111 1711 201112 686

In [0]:
display(customers_count_df.orderBy("YYYYMM"))

YYYYMM active_users 200912 1045 201001 786 201002 807 201003 1111 201004 998 201005 1062 201006 1095 201007 988 201008 964 201009 1202 201010 1577 201011 1683 201012 948 201101 783 201102 798 201103 1020 201104 899 201105 1079 201106 1051 201107 993 201108 980 201109 1302 201110 1425 201111 1711 201112 686

## New VS Exisiting Users per Month
- New users are customer_ids that appear for the first time in the whole dataset
- Existing users are customer_ids that have already appeared in previous months 
- We can reuse customers_df from the previous query since it contains all the customer_ids for each month

In [0]:
display(customers_df.orderBy("YYYYMM"))

YYYYMM customer_id 200912 17804 200912 15998 200912 17804 200912 17592 200912 15998 200912 16393 200912 17804 200912 12636 200912 17592 200912 16329 200912 15998 200912 17984 200912 16393 200912 14040 200912 17804 200912 18102 200912 14110 200912 16321 200912 17592 200912 16011 200912 16329 200912 14911 200912 15998 200912 15998 200912 17984 200912 13394 200912 16393 200912 16393 200912 14040 200912 14341 200912 17804 200912 13078 200912 18102 200912 18087 200912 14110 200912 13758 200912 16321 200912 16167 200912 17865 200912 13293 200912 16011 200912 15311 200912 16329 200912 17700 200912 14911 200912 15998 200912 15998 200912 15998 200912 15998 200912 12533 200912 17984 200912 17984 200912 13394 200912 16011 200912 16393 200912 16393 200912 16393 200912 16393 200912 14040 200912 14040 200912 14341 200912 17804 200912 17804 200912 13078 200912 13078 200912 15362 200912 18102 200912 12682 200912 18087 200912 13635 200912 14110 200912 17519 200912 13758 200912 13758 200912 16321 200912 17592 200912 16167 200912 17865 200912 17865 200912 13767 200912 null 200912 17238 200912 17238 200912 15311 200912 15311 200912 16329 200912 16329 200912 16329 200912 17700 200912 17700 200912 14911 200912 14911 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 12533 200912 12533 200912 12533 200912 17984 200912 17984 200912 17984 200912 14871 200912 13394 200912 16011 200912 16011 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 14040 200912 14040 200912 14040 200912 14040 200912 15061 200912 14292 200912 17804 200912 17804 200912 17804 200912 17804 200912 13085 200912 13078 200912 13078 200912 13078 200912 15362 200912 15362 200912 15362 200912 18102 200912 18102 200912 12682 200912 12682 200912 18087 200912 13635 200912 13635 200912 13635 200912 14110 200912 17519 200912 17519 200912 13758 200912 13758 200912 13758 200912 13758 200912 16321 200912 16321 200912 17592 200912 17592 200912 16167 200912 16167 200912 17865 200912 17865 200912 17865 200912 17865 200912 17592 200912 13767 200912 13767 200912 null 200912 17238 200912 17238 200912 17238 200912 17238 200912 15712 200912 15311 200912 15311 200912 15311 200912 15311 200912 16329 200912 16329 200912 16329 200912 16329 200912 16329 200912 17700 200912 17700 200912 17700 200912 17700 200912 17700 200912 14911 200912 14911 200912 14911 200912 14911 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 15998 200912 12533 200912 12533 200912 12533 200912 12533 200912 12533 200912 12533 200912 16714 200912 17984 200912 17984 200912 17984 200912 17984 200912 17984 200912 17984 200912 14871 200912 14871 200912 13394 200912 13394 200912 16011 200912 16011 200912 16011 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 16393 200912 14040 200912 14040 200912 14040 200912 14040 200912 14040 200912 14040 200912 14040 200912 12913 200912 15061 200912 15061 200912 15061 200912 14341 200912 17804 200912 17804 200912 17804 200912 17804 200912 17804 200912 17804 200912 17804 200912 13085 200912 13085 200912 13085 200912 13078 200912 13078 200912 13078 200912 13078 200912 13078 200912 15362 200912 15362 200912 15362 200912 15362 200912 15362 200912 15362 200912 18102 200912 18102 200912 18102 200912 18102 200912 12682 200912 12682 200912 12682 200912 12682 200912 12682 200912 18087 200912 13635 200912 13635 200912 13635 200912 13635 200912 13635 200912 13635 200912 13635 200912 14110 200912 17519 200912 17519 200912 17519 200912 17519 200912 17519 200912 13758 200912 13758 200912 13758 200912 13758 200912 13758 200912 13758 200912 13758 200912 15413 200912 16321 200912 16321 200912 16321 200912 16321 200912 16321 200912 17592 20091

#### New users: dropping duplicate customer_ids and only keeping the first copy gives us the new users for each month

In [0]:
new_users_df = customers_df.dropDuplicates(["customer_id"])

#### Old users: substracting new_users_df from the original customers_df gives us the existing users for each month

In [0]:
old_users_df = customers_df.subtract(new_users_df)

#### Count the number of new/existing users for each month

In [0]:
new_users_count_df = new_users_df.groupBy("YYYYMM").count()
new_users_count_df = new_users_count_df.withColumnRenamed("count", "new_users")
display(new_users_count_df.orderBy("YYYYMM"))

YYYYMM new_users 200912 1046 201001 394 201002 363 201003 436 201004 291 201005 254 201006 269 201007 183 201008 158 201009 242 201010 379 201011 322 201012 77 201101 71 201102 123 201103 178 201104 105 201105 108 201106 108 201107 102 201108 106 201109 187 201110 221 201111 192 201112 28

In [0]:
old_users_count_df = old_users_df.groupBy("YYYYMM").count()
old_users_count_df = old_users_count_df.withColumnRenamed("count", "existing_users")
display(old_users_count_df.orderBy("YYYYMM"))

YYYYMM existing_users 201001 393 201002 445 201003 676 201004 708 201005 809 201006 827 201007 806 201008 807 201009 961 201010 1199 201011 1362 201012 872 201101 713 201102 676 201103 843 201104 795 201105 972 201106 944 201107 892 201108 875 201109 1116 201110 1205 201111 1520 201112 659

#### Merge new and existing users count for each month

In [0]:
new_old_users_count_df = new_users_count_df.join(old_users_count_df, "YYYYMM", "left_outer")
display(new_old_users_count_df.orderBy("YYYYMM"))

YYYYMM new_users existing_users 200912 1046 null 201001 394 393 201002 363 445 201003 436 676 201004 291 708 201005 254 809 201006 269 827 201007 183 806 201008 158 807 201009 242 961 201010 379 1199 201011 322 1362 201012 77 872 201101 71 713 201102 123 676 201103 178 843 201104 105 795 201105 108 972 201106 108 944 201107 102 892 201108 106 875 201109 187 1116 201110 221 1205 201111 192 1520 201112 28 659

In [0]:
display(new_old_users_count_df.orderBy("YYYYMM"))

YYYYMM new_users existing_users 200912 1046 null 201001 394 393 201002 363 445 201003 436 676 201004 291 708 201005 254 809 201006 269 827 201007 183 806 201008 158 807 201009 242 961 201010 379 1199 201011 322 1362 201012 77 872 201101 71 713 201102 123 676 201103 178 843 201104 105 795 201105 108 972 201106 108 944 201107 102 892 201108 106 875 201109 187 1116 201110 221 1205 201111 192 1520 201112 28 659

## RFM Segmentation
RFM Segmentation is a segmentation technique that helps determine marketing and sales strategies based on buying habits of customers.
- RFM = Recency, Frequency, Monetary
- Recency: Time since customer's last purchase
- Frequency: Total number of purchases
- Monetary: Total spending by the customer
- Segment customers into meaningful groups depending on their values for Recency, Frequency, and Monetary
- This segmentation is used to gain insights into customer value and develop targeted marketting strategies for important segments

In [0]:
display(retail_df)

invoice_no stock_code description quantity invoice_date unit_price customer_id country 489434 85048 15CM CHRISTMAS GLASS BALL 20 LIGHTS 12 2009-12-01T07:45:00.000+0000 6.95 13085 United Kingdom 489434 79323P PINK CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 79323W WHITE CHERRY LIGHTS 12 2009-12-01T07:45:00.000+0000 6.75 13085 United Kingdom 489434 22041 RECORD FRAME 7" SINGLE SIZE 48 2009-12-01T07:45:00.000+0000 2.1 13085 United Kingdom 489434 21232 STRAWBERRY CERAMIC TRINKET BOX 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 22064 PINK DOUGHNUT TRINKET POT 24 2009-12-01T07:45:00.000+0000 1.65 13085 United Kingdom 489434 21871 SAVE THE PLANET MUG 24 2009-12-01T07:45:00.000+0000 1.25 13085 United Kingdom 489434 21523 FANCY FONT HOME SWEET HOME DOORMAT 10 2009-12-01T07:45:00.000+0000 5.95 13085 United Kingdom 489435 22350 CAT BOWL 12 2009-12-01T07:46:00.000+0000 2.55 13085 United Kingdom 489435 22349 DOG BOWL , CHASING BALL DESIGN 12 2009-12-01T07:46:00.000+0000 3.75 13085 United Kingdom 489435 22195 HEART MEASURING SPOONS LARGE 24 2009-12-01T07:46:00.000+0000 1.65 13085 United Kingdom 489435 22353 LUNCHBOX WITH CUTLERY FAIRY CAKES 12 2009-12-01T07:46:00.000+0000 2.55 13085 United Kingdom 489436 48173C DOOR MAT BLACK FLOCK 10 2009-12-01T09:06:00.000+0000 5.95 13078 United Kingdom 489436 21755 LOVE BUILDING BLOCK WORD 18 2009-12-01T09:06:00.000+0000 5.45 13078 United Kingdom 489436 21754 HOME BUILDING BLOCK WORD 3 2009-12-01T09:06:00.000+0000 5.95 13078 United Kingdom 489436 84879 ASSORTED COLOUR BIRD ORNAMENT 16 2009-12-01T09:06:00.000+0000 1.69 13078 United Kingdom 489436 22119 PEACE WOODEN BLOCK LETTERS 3 2009-12-01T09:06:00.000+0000 6.95 13078 United Kingdom 489436 22142 CHRISTMAS CRAFT WHITE FAIRY 12 2009-12-01T09:06:00.000+0000 1.45 13078 United Kingdom 489436 22296 HEART IVORY TRELLIS LARGE 12 2009-12-01T09:06:00.000+0000 1.65 13078 United Kingdom 489436 22295 HEART FILIGREE DOVE LARGE 12 2009-12-01T09:06:00.000+0000 1.65 13078 United Kingdom 489436 22109 FULL ENGLISH BREAKFAST PLATE 16 2009-12-01T09:06:00.000+0000 3.39 13078 United Kingdom 489436 22107 PIZZA PLATE IN BOX 4 2009-12-01T09:06:00.000+0000 3.75 13078 United Kingdom 489436 22194 BLACK DINER WALL CLOCK 2 2009-12-01T09:06:00.000+0000 8.5 13078 United Kingdom 489436 35004B SET OF 3 BLACK FLYING DUCKS 12 2009-12-01T09:06:00.000+0000 4.65 13078 United Kingdom 489436 82582 AREA PATROLLED METAL SIGN 12 2009-12-01T09:06:00.000+0000 2.1 13078 United Kingdom 489436 21181 PLEASE ONE PERSON METAL SIGN 12 2009-12-01T09:06:00.000+0000 2.1 13078 United Kingdom 489436 21756 BATH BUILDING BLOCK WORD 3 2009-12-01T09:06:00.000+0000 5.95 13078 United Kingdom 489436 21333 CLASSIC WHITE FRAME 6 2009-12-01T09:06:00.000+0000 2.95 13078 United Kingdom 489436 84596F SMALL MARSHMALLOWS PINK BOWL 8 2009-12-01T09:06:00.000+0000 1.25 13078 United Kingdom 489436 84596L BISCUITS SMALL BOWL LIGHT BLUE 8 2009-12-01T09:06:00.000+0000 1.25 13078 United Kingdom 489436 22111 SCOTTIE DOG HOT WATER BOTTLE 24 2009-12-01T09:06:00.000+0000 4.25 13078 United Kingdom 489437 22143 CHRISTMAS CRAFT HEART DECORATIONS 6 2009-12-01T09:08:00.000+0000 2.1 15362 United Kingdom 489437 22145 CHRISTMAS CRAFT HEART STOCKING 6 2009-12-01T09:08:00.000+0000 2.1 15362 United Kingdom 489437 22130 PARTY CONE CHRISTMAS DECORATION 12 2009-12-01T09:08:00.000+0000 0.85 15362 United Kingdom 489437 21364 PEACE SMALL WOOD LETTERS 2 2009-12-01T09:08:00.000+0000 6.75 15362 United Kingdom 489437 21360 JOY LARGE WOOD LETTERS 1 2009-12-01T09:08:00.000+0000 9.95 15362 United Kingdom 489437 21351 CINAMMON & ORANGE WREATH 2 2009-12-01T09:08:00.000+0000 6.75 15362 United Kingdom 489437 21352 EUCALYPTUS & PINECONE WREATH 2 2009-12-01T09:08:00.000+0000 6.75 15362 United Kingdom 489437 35400 WOODEN BOX ADVENT CALENDAR 2 2009-12-01T09:08:00.000+0000 8.95 15362 United Kingdom 489437 20695 FLORAL BLUE MONSTER 3 2009-12-01T09:08:00.000+0000 4.25 15362 United Kingdom 489437 37370 RETRO COFFEE MUG

#### Change the invoice_date type to date and make a new column with a literal date of 'January 1, 2012' to facilitate the Recency measure; 
#### Make a column with amount = (unit_price * quantity) to faciliate Monetary measure

In [0]:
rfm_tmp = retail_df.select("invoice_no", "quantity", "unit_price", "customer_id", to_date(retail_df.invoice_date).alias("invoice_date"), to_date(lit("2012-01-01")).alias("today"), 
                           expr("ROUND(unit_price * quantity) AS total_price"))
display(rfm_tmp)

invoice_no quantity unit_price customer_id invoice_date today total_price 489434 12 6.95 13085 2009-12-01 2012-01-01 83.0 489434 12 6.75 13085 2009-12-01 2012-01-01 81.0 489434 12 6.75 13085 2009-12-01 2012-01-01 81.0 489434 48 2.1 13085 2009-12-01 2012-01-01 101.0 489434 24 1.25 13085 2009-12-01 2012-01-01 30.0 489434 24 1.65 13085 2009-12-01 2012-01-01 40.0 489434 24 1.25 13085 2009-12-01 2012-01-01 30.0 489434 10 5.95 13085 2009-12-01 2012-01-01 60.0 489435 12 2.55 13085 2009-12-01 2012-01-01 31.0 489435 12 3.75 13085 2009-12-01 2012-01-01 45.0 489435 24 1.65 13085 2009-12-01 2012-01-01 40.0 489435 12 2.55 13085 2009-12-01 2012-01-01 31.0 489436 10 5.95 13078 2009-12-01 2012-01-01 60.0 489436 18 5.45 13078 2009-12-01 2012-01-01 98.0 489436 3 5.95 13078 2009-12-01 2012-01-01 18.0 489436 16 1.69 13078 2009-12-01 2012-01-01 27.0 489436 3 6.95 13078 2009-12-01 2012-01-01 21.0 489436 12 1.45 13078 2009-12-01 2012-01-01 17.0 489436 12 1.65 13078 2009-12-01 2012-01-01 20.0 489436 12 1.65 13078 2009-12-01 2012-01-01 20.0 489436 16 3.39 13078 2009-12-01 2012-01-01 54.0 489436 4 3.75 13078 2009-12-01 2012-01-01 15.0 489436 2 8.5 13078 2009-12-01 2012-01-01 17.0 489436 12 4.65 13078 2009-12-01 2012-01-01 56.0 489436 12 2.1 13078 2009-12-01 2012-01-01 25.0 489436 12 2.1 13078 2009-12-01 2012-01-01 25.0 489436 3 5.95 13078 2009-12-01 2012-01-01 18.0 489436 6 2.95 13078 2009-12-01 2012-01-01 18.0 489436 8 1.25 13078 2009-12-01 2012-01-01 10.0 489436 8 1.25 13078 2009-12-01 2012-01-01 10.0 489436 24 4.25 13078 2009-12-01 2012-01-01 102.0 489437 6 2.1 15362 2009-12-01 2012-01-01 13.0 489437 6 2.1 15362 2009-12-01 2012-01-01 13.0 489437 12 0.85 15362 2009-12-01 2012-01-01 10.0 489437 2 6.75 15362 2009-12-01 2012-01-01 14.0 489437 1 9.95 15362 2009-12-01 2012-01-01 10.0 489437 2 6.75 15362 2009-12-01 2012-01-01 14.0 489437 2 6.75 15362 2009-12-01 2012-01-01 14.0 489437 2 8.95 15362 2009-12-01 2012-01-01 18.0 489437 3 4.25 15362 2009-12-01 2012-01-01 13.0 489437 12 1.25 15362 2009-12-01 2012-01-01 15.0 489437 12 0.85 15362 2009-12-01 2012-01-01 10.0 489437 6 2.55 15362 2009-12-01 2012-01-01 15.0 489437 3 4.25 15362 2009-12-01 2012-01-01 13.0 489437 12 0.65 15362 2009-12-01 2012-01-01 8.0 489437 12 0.85 15362 2009-12-01 2012-01-01 10.0 489437 12 0.85 15362 2009-12-01 2012-01-01 10.0 489437 12 1.25 15362 2009-12-01 2012-01-01 15.0 489437 6 2.95 15362 2009-12-01 2012-01-01 18.0 489437 6 2.95 15362 2009-12-01 2012-01-01 18.0 489437 6 2.95 15362 2009-12-01 2012-01-01 18.0 489437 4 3.75 15362 2009-12-01 2012-01-01 15.0 489437 3 4.95 15362 2009-12-01 2012-01-01 15.0 489437 3 4.95 15362 2009-12-01 2012-01-01 15.0 489438 28 0.98 18102 2009-12-01 2012-01-01 27.0 489438 30 1.69 18102 2009-12-01 2012-01-01 51.0 489438 30 1.15 18102 2009-12-01 2012-01-01 35.0 489438 30 2.0 18102 2009-12-01 2012-01-01 60.0 489438 60 1.3 18102 2009-12-01 2012-01-01 78.0 489438 32 2.5 18102 2009-12-01 2012-01-01 80.0 489438 32 2.5 18102 2009-12-01 2012-01-01 80.0 489438 56 3.0 18102 2009-12-01 2012-01-01 168.0 489438 56 3.0 18102 2009-12-01 2012-01-01 168.0 489438 60 1.9 18102 2009-12-01 2012-01-01 114.0 489438 56 1.9 18102 2009-12-01 2012-01-01 106.0 489438 56 2.15 18102 2009-12-01 2012-01-01 120.0 489438 60 2.4 18102 2009-12-01 2012-01-01 144.0 489438 60 6.38 18102 2009-12-01 2012-01-01 383.0 489438 60 6.4 18102 2009-12-01 2012-01-01 384.0 489438 60 2.4 18102 2009-12-01 2012-01-01 144.0 489438 60 2.4 18102 2009-12-01 2012-01-01 144.0 489439 12 1.45 12682 2009-12-01 2012-01-01 17.0 489439 9 4.95 12682 2009-12-01 2012-01-01 45.0 489439 9 4.95 12682 2009-12-01 2012-01-01 45.0 489439 12 2.55 12682 2009-12-01 2012-01-01 31.0 489439 3 5.95 12682 2009-12-01 2012-01-01 18.0 489439 3 5.95 12682 2009-12-01 2012-01-01 18.0 489439 25 0.42 12682 2009-12-01 2012-01-01 11.0 489439 25 0.42 12682 2009-12-01 2012-01-01 11.0 489439 6 1.95 12682 2009-12-01 2012-01-01 12.0 489439 8 1.65 12682 2009-12-01 2012-01-01 13.0 489439 12 0.65 12682 2009-12-01 2012-01-01 8.0 489439 12 0.85 12682 2009-12

#### Remove entries with negavtive values and drop missing values to faciliate evaluation

In [0]:
rfm_tmp = rfm_tmp.filter(rfm_tmp.quantity > 0).filter(rfm_tmp.unit_price > 0).filter(rfm_tmp.total_price > 0)
rfm_tmp = rfm_tmp.na.drop()
rfm_tmp.count()

Out[29]: 800167

#### Find the Recency and Monetary values

In [0]:
df_tmp = rfm_tmp.groupBy("customer_id").agg(sum("total_price").alias("sum_price"), max("invoice_date").alias("max_date"), max("today").alias("today"))
df_x = df_tmp.select("customer_id", "sum_price", datediff(col("today"), col("max_date")).alias("recency"))
display(df_x)

customer_id sum_price recency 13623 2596.0 53 17679 3579.0 75 17389 57252.0 23 18051 2283.0 657 13289 310.0 746 17753 392.0 487 15727 9535.0 39 16574 1298.0 94 14832 323.0 653 15447 496.0 353 13285 3408.0 46 14450 1133.0 203 16503 3642.0 129 12799 221.0 552 16861 949.0 82 14570 617.0 303 13832 617.0 42 15957 936.0 54 13840 658.0 439 17420 1097.0 73 16386 1111.0 51 15846 110.0 408 16339 111.0 307 18024 391.0 175 12940 923.0 77 15619 337.0 33 15790 225.0 33 15967 3099.0 46 12471 39995.0 25 17172 954.0 101 14514 4623.0 84 18201 452.0 745 15100 2894.0 356 14465 3947.0 269 13622 662.0 719 14423 917.0 241 15003 1090.0 496 15538 3360.0 560 18218 1145.0 234 14876 258.0 544 15254 1097.0 150 16500 843.0 27 14075 1737.0 27 16916 1141.0 46 14148 467.0 256 17688 306.0 416 18161 1812.0 38 17607 337.0 399 12626 6632.0 46 13261 945.0 291 14536 175.0 282 15004 1265.0 170 18282 181.0 30 12715 360.0 129 12611 1193.0 75 12367 170.0 27 13648 1775.0 423 17809 11347.0 39 16791 4099.0 52 17044 11969.0 48 15162 315.0 754 15738 13392.0 41 17754 4599.0 23 13875 4858.0 76 17028 1608.0 604 14324 705.0 432 14420 1641.0 122 14714 914.0 47 15296 10397.0 73 14837 3463.0 112 13009 236.0 695 14997 778.0 69 12998 923.0 515 16283 4462.0 28 16534 635.0 665 15173 1513.0 403 13207 534.0 38 17077 306.0 601 15207 681.0 439 13601 1198.0 68 16680 825.0 66 17783 387.0 95 17008 81.0 452 12393 2413.0 95 17708 893.0 210 17751 162.0 444 18221 1430.0 45 14289 919.0 31 14377 617.0 214 17775 56.0 277 13060 307.0 276 15655 562.0 266 17837 1606.0 65 16924 857.0 44 13483 140.0 88 17193 2314.0 58 14958 963.0 170 15382 13778.0 37 13916 1283.0 486 13508 289.0 266 14315 3220.0 67 12366 500.0 656 14809 688.0 644 15575 30.0 642 13465 929.0 469 17437 372.0 552 13898 686.0 348 16224 1620.0 80 17223 887.0 333 18043 884.0 47 16828 129.0 116 15604 218.0 75 17712 524.0 67 13188 115.0 34 14713 5484.0 32 13329 5102.0 54 13225 16267.0 26 17950 1449.0 95 17113 2348.0 597 14944 9784.0 52 13765 355.0 676 14770 3652.0 255 17190 1758.0 81 13795 836.0 411 16094 3698.0 34 16117 661.0 212 16565 775.0 387 16232 2717.0 66 15535 1097.0 196 13398 674.0 215 15960 416.0 397 14779 390.0 303 15292 186.0 224 16706 534.0 60 14660 290.0 109 17154 331.0 98 15429 405.0 49 16999 486.0 46 13504 300.0 87 12607 1585.0 83 13923 350.0 76 15740 155.0 72 17836 109.0 70 14752 390.0 66 13468 13466.0 24 16985 10967.0 40 15070 848.0 395 16119 1723.0 31 16027 2047.0 114 16510 2107.0 395 14365 644.0 417 17438 1433.0 276 15221 2468.0 389 16549 13374.0 33 17346 5376.0 26 16526 721.0 194 13997 388.0 542 14874 116.0 558 15026 1076.0 103 15462 740.0 117 14846 438.0 450 12694 1847.0 93 15647 170.0 438 13500 1469.0 46 17036 179.0 406 14458 243.0 403 15146 1331.0 187 15594 1783.0 38 14124 419.0 107 14054 95.0 88 16295 1336.0 32 16506 90.0 42 13313 3474.0 45 15983 3976.0 45 14030 8359.0 41 17213 5218.0 63 17677 36721.0 24 14477 1970.0 42 15827 3810.0 96 16255 4946.0 39 15584 1291.0 46 14284 1310.0 53 13544 844.0 62 12530 3781.0 82 15558 216.0 662 16601 686.0 63 15216 487.0 110 14812 185.0 621 12717 1315.0 174 16391 1125.0 423 15931 240.0 451 16724 149.0 427 17903 384.0 417 17782 679.0 405 17378 11.0 397 15781 347.0 48 17613 4589.0 45 14245 1708.0 242 16257 47.0 199 16505 779.0 54 17904 195.0 44 14771 593.0 644 15633 4382.0 532 15157 2646.0 26 14532 3893.0 259 16954 6073.0 23 14556 843.0 619 12727 5669.0 33 16947 960.0 258 16741 176.0 648 18174 163.0 30 16835 1553.0 151 15966 433.0 473 17392 633.0 329 13701 161.0 469 17396 8809.0 62 17924 3539.0 34 15161 119.0 411 15634 560.0 40 12559 565.0 333 14760 90.0 320 17462 680.0 83 16766 1150.0 296 12757 7901.0 38 15348 159.0 95 12723 733.0 26 17502 690.0 32 16800 1215.0 34 17969 2404.0 189 12523 4399.0 42 15270 8537.0 38 15164 2101.0 91 15091 1051.0 90 17774 3374.0 119 13214 3287.0 440 16648 673.0 65 16317 2206.0 142 14437 960.0 388 17926 1042.0 156 16234 333.0 499 14058 473.0 485 18107 169.0 465 13022 66.0 452 15973 987.0 389 17525 1166.0 382 15968 423.0 411 12401 85.0 326 17972 137.0 264 17265 

#### Find the Frequency value

In [0]:
df_y = rfm_tmp.groupBy("customer_id", "invoice_no").agg(sum("total_price").alias("sum_price2"))
df_z = df_y.groupBy("customer_id").agg(count("sum_price2").alias("sum_price2"))
display(df_z)

customer_id sum_price2 16574 3 15727 15 17389 61 15447 5 15619 1 18051 7 14450 7 15846 1 15790 1 13285 6 17753 4 14570 3 14832 1 13832 2 17679 7 13623 10 16861 5 15957 3 16503 11 13840 1 13289 1 17420 6 16386 3 18024 2 12799 1 16339 1 12940 2 15254 3 12471 79 15538 6 14465 11 15967 12 14423 3 18161 6 14536 1 12626 9 15003 6 18218 4 16500 9 15100 7 14148 2 14075 6 18201 1 18282 2 13261 2 14514 9 16916 4 15004 1 13622 1 17172 3 17607 1 12715 1 14876 2 17688 1 12367 1 12611 1 15738 41 13875 14 17044 12 17077 1 16283 8 13207 4 14377 2 17837 8 16791 15 17809 20 13601 4 15173 7 14289 6 17754 13 16680 3 15207 3 13648 8 15655 1 15296 12 14837 9 18221 6 14714 4 12393 6 16924 3 17028 3 14420 4 17783 3 17008 1 14997 4 13060 1 12998 3 17775 1 17708 4 14324 2 13009 1 13483 1 16534 1 15162 1 17751 1 16224 5 17193 9 17223 3 15382 12 13898 2 13508 2 13916 4 14315 8 15575 1 15604 1 13465 2 17437 1 14809 2 14958 6 16828 2 13188 1 18043 3 17712 1 12366 1 14770 8 14713 21 13329 8 14944 19 17113 6 16094 11 15292 1 15960 4 17190 7 13225 16 14779 3 15535 6 16999 2 16232 7 17950 7 16117 3 16706 3 13795 4 15740 1 15429 2 16565 4 13923 1 13504 1 14660 2 17836 1 17154 1 12607 1 13398 2 14752 1 13765 1 16549 34 15462 4 16985 23 13468 72 16027 15 17346 26 15221 4 17438 5 13997 3 16510 6 12694 3 15070 3 15026 5 14846 1 15146 2 16295 4 16526 2 14874 1 16506 2 14365 2 16119 8 13500 4 15594 5 14124 1 17036 1 15647 1 14458 1 14054 1 14284 3 13313 12 12530 9 17677 54 16255 15 14030 20 15827 28 17613 14 16391 4 17782 2 17378 1 15983 8 17903 1 12717 3 17213 13 14477 7 15584 6 15931 1 15781 3 14245 2 16724 1 16601 3 15216 2 13544 6 17904 1 16505 1 16257 1 15558 1 14812 1 15157 5 14556 3 15966 1 17396 14 12727 10 14532 11 12723 5 15634 2 15633 13 12757 13 16766 2 16947 5 13701 1 14760 1 16954 20 17924 8 16835 6 16800 1 14771 2 17502 4 12559 2 15348 1 16741 1 17392 3 17462 2 18174 2 15161 1 17774 9 12401 1 12436 1 17969 20 12523 23 15270 21 17972 1 15973 3 16234 3 15904 1 17926 5 13214 7 17525 3 15164 9 14437 5 18014 1 16317 5 15091 3 17265 2 14058 2 15968 1 16648 3 13022 1 18107 1 13288 5 14085 33 16032 6 17225 1 15191 3 13494 1 17639 6 13003 3 17430 4 16813 10 13516 4 13172 1 17554 5 13046 3 16544 4 15491 6 13915 1 17253 1 14658 2 17067 3 15658 8 13967 2 14064 15 15093 23 13720 4 12943 2 17588 7 15497 9 17388 11 15202 3 15557 1 14332 5 13994 6 16801 4 16968 2 13751 1 16913 3 12384 2 12542 2 16798 3 14801 1 15962 1 17074 1 16738 1 15470 1 14589 1 14715 8 13594 2 15850 6 12494 9 13368 4 18030 2 12755 11 17508 8 14336 4 15716 11 18193 3 15902 1 14935 2 16819 2 12782 11 17792 2 17397 3 16125 14 16274 1 16375 2 12796 1 12854 7 16605 1 12355 2 15637 5 18105 1 18085 4 13170 1 13066 2 17347 1 12837 2 12566 1 16721 1 17740 6 14101 13 13903 1 18258 6 14044 10 13782 13 12362 11 17353 4 16930 3 17001 19 16002 1 13220 5 14410 11 15844 1 16707 4 15116 4 16189 2 17798 1 16532 5 15479 1 17196 1 15130 1 13960 2 13377 2 16671 3 14501 2 13745 2 14703 1 13830 1 14999 1 14525 17 14180 33 18071 6 15398 2 14558 3 15795 8 13027 19 17059 9 14555 6 17678 2 14554 4 13109 3 18240 1 16244 2 15840 2 13904 2 15060 4 17808 4 12447 1 17297 1 17981 1 17647 3 15253 8 14402 1 13726 2 16839 44 14733 15 14367 16 13045 3 17741 2 16455 11 16265 12 14776 15 14056 23 17955 6 17081 6 16366 4 14814 1 17147 9 16314 3 14160 7 14283 4 15234 1 12961 1 17890 2 13775 1 13195 1 18278 2 14687 3 13665 3 15961 1 15503 2 14117 1 15096 1 14642 1 13104 9 12948 9 16043 8 12453 1 15105 9 13166 2 16053 6 12373 3 14177 13 14141 4 12885 3 12913 13 17281 6 17542 3 17207 1 13738 1 16944 1 13371 2 13836 2 13873 8 17983 1 15666 2 15923 1 15733 1 14647 2 13415 4 14302 3 12548 1 13676 2 15260 25 15531 2 16684 55 15062 8 14496 7 16012 5 15024 16 15242 1 13139 8 16774 8 14419 7 15219 2 13052 7 12725 4 15332 6 16175 7 15855 6 12914 1 17463 7 16699 7 14455 4 17148 1 12711 2 16331 6 18248 2 18113 1 15262 3 16264 2 17070 2 15138 1 15588 1 15404 1 15583 1 15073 8 15426 9 12582 4 14062 24 13523 6 16326 24 16553 33 17637 2 12705 14 17244 5

#### Create the RFM table
The table can be read as: the customer with the customer_id 12346 has spent a total of £77,551; the last purchase he made was about 1 year ago (348 days); and the number of purchases he has made is 12.

In [0]:
rfm_df = df_x.join(df_z, "customer_id", "inner")
rfm_df = rfm_df.withColumnRenamed("sum_price", "monetary").withColumnRenamed("sum_price2", "frequency")
display(rfm_df.orderBy("customer_id"))

customer_id monetary recency frequency 12346 77551.0 348 12 12347 5641.0 25 8 12348 2016.0 98 5 12349 4454.0 41 4 12350 334.0 333 1 12351 304.0 398 1 12352 2855.0 59 10 12353 409.0 227 2 12354 1086.0 255 1 12355 950.0 237 2 12356 6371.0 45 6 12357 18287.0 56 3 12358 3893.0 24 5 12359 8990.0 80 10 12360 4267.0 75 8 12361 516.0 310 4 12362 5381.0 26 11 12363 549.0 132 2 12364 1315.0 30 4 12365 644.0 314 2 12366 500.0 656 1 12367 170.0 27 1 12368 920.0 651 1 12369 1800.0 436 3 12370 4339.0 74 7 12371 4083.0 67 5 12372 1302.0 94 3 12373 1137.0 334 3 12374 2992.0 48 4 12375 461.0 33 2 12376 504.0 412 2 12377 3428.0 338 4 12378 5440.0 152 2 12379 1631.0 104 4 12380 9727.0 44 11 12381 1850.0 27 5 12383 1856.0 207 5 12384 589.0 51 2 12385 1950.0 402 1 12386 661.0 360 2 12387 143.0 438 1 12388 3903.0 38 8 12389 1436.0 425 3 12390 554.0 102 1 12391 2799.0 44 6 12392 236.0 614 1 12393 2413.0 95 6 12394 1274.0 86 2 12395 5080.0 42 15 12396 937.0 697 1 12397 2413.0 58 2 12398 1639.0 68 1 12399 1959.0 142 7 12400 205.0 437 1 12401 85.0 326 1 12402 1271.0 346 6 12403 1020.0 72 2 12404 63.0 705 1 12405 1714.0 171 1 12406 6394.0 45 5 12407 1712.0 72 5 12408 4356.0 55 8 12409 23426.0 101 7 12410 1016.0 331 3 12411 354.0 432 1 12412 2595.0 97 6 12413 1008.0 89 4 12414 954.0 240 6 12415 144470.0 47 28 12416 202.0 679 1 12417 6857.0 26 20 12418 2027.0 135 4 12419 505.0 408 2 12420 604.0 86 1 12421 1101.0 38 6 12422 5258.0 118 17 12423 2640.0 23 10 12424 3346.0 185 4 12425 1752.0 101 2 12426 1354.0 217 2 12427 1131.0 44 3 12428 7992.0 48 9 12429 7488.0 32 8 12430 256.0 66 1 12431 10905.0 58 27 12432 5069.0 65 5 12433 20577.0 23 10 12434 1113.0 107 3 12435 13931.0 102 6 12436 511.0 122 1 12437 12702.0 24 39 12438 2909.0 37 2 12439 1164.0 705 2 12440 864.0 485 2 12441 352.0 389 1 12442 172.0 26 1 12443 487.0 755 1 12444 5024.0 44 5 12445 437.0 45 2 12446 1014.0 80 1 12447 479.0 266 1 12448 451.0 67 1 12449 4841.0 45 6 12450 693.0 179 5 12451 12708.0 33 8 12452 432.0 39 2 12453 712.0 157 1 12454 16681.0 79 5 12455 5900.0 96 10 12456 3189.0 67 3 12457 3000.0 89 8 12458 1790.0 94 4 12459 453.0 628 1 12460 329.0 479 1 12461 832.0 117 2 12462 1197.0 41 2 12463 1350.0 76 4 12464 1215.0 33 8 12465 744.0 40 2 12466 57.0 705 1 12467 135.0 446 1 12468 3681.0 210 5 12469 3082.0 545 2 12470 215.0 712 1 12471 39995.0 25 79 12472 17947.0 55 19 12473 3955.0 52 5 12474 12479.0 40 31 12475 1157.0 76 2 12476 14165.0 38 23 12477 20333.0 45 9 12478 686.0 26 1 12479 527.0 34 1 12480 5491.0 51 7 12481 12379.0 45 15 12482 23686.0 599 29 12483 5126.0 40 19 12484 6421.0 55 12 12485 711.0 552 1 12486 1041.0 454 1 12487 286.0 753 1 12488 1305.0 32 3 12489 337.0 359 1 12490 9878.0 28 21 12491 465.0 61 1 12492 721.0 86 3 12493 440.0 188 2 12494 1888.0 38 9 12495 356.0 663 1 12496 54.0 643 2 12497 6255.0 104 3 12498 409.0 44 3 12499 399.0 583 1 12500 5973.0 46 19 12501 2197.0 359 1 12502 3743.0 118 5 12503 1126.0 607 1 12504 779.0 41 4 12505 960.0 431 1 12506 73.0 255 1 12507 1842.0 157 4 12508 397.0 49 2 12509 179.0 307 1 12510 5189.0 165 9 12511 3096.0 451 1 12512 140.0 89 2 12513 618.0 135 1 12514 1605.0 290 2 12515 384.0 376 1 12516 2650.0 100 4 12517 2518.0 51 5 12518 2064.0 23 5 12519 729.0 86 2 12520 2642.0 102 5 12521 602.0 165 1 12522 193.0 62 2 12523 4399.0 42 23 12524 11149.0 32 17 12525 985.0 521 3 12526 1324.0 23 3 12527 735.0 104 5 12528 1940.0 32 6 12529 319.0 551 1 12530 3781.0 82 9 12531 1946.0 37 2 12532 351.0 53 1 12533 1374.0 650 2 12534 1098.0 153 1 12535 720.0 114 2 12536 12621.0 66 3 12537 731.0 431 1 12538 644.0 52 2 12539 11270.0 45 7 12540 22295.0 42 33 12541 981.0 52 3 12542 483.0 507 2 12543 981.0 471 1 12544 373.0 59 1 12545 1001.0 98 2 12546 1171.0 69 2 12547 381.0 220 3 12548 94.0 189 1 12549 1576.0 598 6 12550 969.0 102 1 12551 173.0 380 1 12552 443.0 61 2 12553 4644.0 31 12 12554 470.0 471 1 12555 129.0 705 1 12556 612.0 73 2 12557 18719.0 39 8 12558 270.0 30 1 12559 565.0 333 2 12560 2139.0 28 3 12561 243.0 325 1 12562 5004.0 31 9 12563 252.0 431

#### RFM scores

In [0]:
from pyspark.ml.feature import QuantileDiscretizer

qds = QuantileDiscretizer(numBuckets = 6, inputCol="recency", outputCol="recency_score")
rfm_df = qds.fit(rfm_df).transform(rfm_df)

qds = QuantileDiscretizer(numBuckets = 6, inputCol="frequency", outputCol="frequency_score")
rfm_df = qds.fit(rfm_df).transform(rfm_df)

qds = QuantileDiscretizer(numBuckets = 6, inputCol = "monetary", outputCol = "monetary_score")
rfm_df = qds.fit(rfm_df).transform(rfm_df)

display(rfm_df.orderBy("customer_id"))

customer_id monetary recency frequency recency_score frequency_score monetary_score 12346 77551.0 348 12 4.0 5.0 5.0 12347 5641.0 25 8 0.0 4.0 5.0 12348 2016.0 98 5 2.0 4.0 4.0 12349 4454.0 41 4 1.0 3.0 5.0 12350 334.0 333 1 4.0 1.0 1.0 12351 304.0 398 1 4.0 1.0 1.0 12352 2855.0 59 10 1.0 5.0 4.0 12353 409.0 227 2 3.0 2.0 1.0 12354 1086.0 255 1 3.0 1.0 3.0 12355 950.0 237 2 3.0 2.0 3.0 12356 6371.0 45 6 1.0 4.0 5.0 12357 18287.0 56 3 1.0 3.0 5.0 12358 3893.0 24 5 0.0 4.0 5.0 12359 8990.0 80 10 2.0 5.0 5.0 12360 4267.0 75 8 2.0 4.0 5.0 12361 516.0 310 4 4.0 3.0 2.0 12362 5381.0 26 11 0.0 5.0 5.0 12363 549.0 132 2 3.0 2.0 2.0 12364 1315.0 30 4 0.0 3.0 3.0 12365 644.0 314 2 4.0 2.0 2.0 12366 500.0 656 1 5.0 1.0 2.0 12367 170.0 27 1 0.0 1.0 0.0 12368 920.0 651 1 5.0 1.0 3.0 12369 1800.0 436 3 4.0 3.0 4.0 12370 4339.0 74 7 2.0 4.0 5.0 12371 4083.0 67 5 2.0 4.0 5.0 12372 1302.0 94 3 2.0 3.0 3.0 12373 1137.0 334 3 4.0 3.0 3.0 12374 2992.0 48 4 1.0 3.0 4.0 12375 461.0 33 2 0.0 2.0 1.0 12376 504.0 412 2 4.0 2.0 2.0 12377 3428.0 338 4 4.0 3.0 4.0 12378 5440.0 152 2 3.0 2.0 5.0 12379 1631.0 104 4 2.0 3.0 3.0 12380 9727.0 44 11 1.0 5.0 5.0 12381 1850.0 27 5 0.0 4.0 4.0 12383 1856.0 207 5 3.0 4.0 4.0 12384 589.0 51 2 1.0 2.0 2.0 12385 1950.0 402 1 4.0 1.0 4.0 12386 661.0 360 2 4.0 2.0 2.0 12387 143.0 438 1 4.0 1.0 0.0 12388 3903.0 38 8 0.0 4.0 5.0 12389 1436.0 425 3 4.0 3.0 3.0 12390 554.0 102 1 2.0 1.0 2.0 12391 2799.0 44 6 1.0 4.0 4.0 12392 236.0 614 1 5.0 1.0 0.0 12393 2413.0 95 6 2.0 4.0 4.0 12394 1274.0 86 2 2.0 2.0 3.0 12395 5080.0 42 15 1.0 5.0 5.0 12396 937.0 697 1 5.0 1.0 3.0 12397 2413.0 58 2 1.0 2.0 4.0 12398 1639.0 68 1 2.0 1.0 3.0 12399 1959.0 142 7 3.0 4.0 4.0 12400 205.0 437 1 4.0 1.0 0.0 12401 85.0 326 1 4.0 1.0 0.0 12402 1271.0 346 6 4.0 4.0 3.0 12403 1020.0 72 2 2.0 2.0 3.0 12404 63.0 705 1 5.0 1.0 0.0 12405 1714.0 171 1 3.0 1.0 4.0 12406 6394.0 45 5 1.0 4.0 5.0 12407 1712.0 72 5 2.0 4.0 4.0 12408 4356.0 55 8 1.0 4.0 5.0 12409 23426.0 101 7 2.0 4.0 5.0 12410 1016.0 331 3 4.0 3.0 3.0 12411 354.0 432 1 4.0 1.0 1.0 12412 2595.0 97 6 2.0 4.0 4.0 12413 1008.0 89 4 2.0 3.0 3.0 12414 954.0 240 6 3.0 4.0 3.0 12415 144470.0 47 28 1.0 5.0 5.0 12416 202.0 679 1 5.0 1.0 0.0 12417 6857.0 26 20 0.0 5.0 5.0 12418 2027.0 135 4 3.0 3.0 4.0 12419 505.0 408 2 4.0 2.0 2.0 12420 604.0 86 1 2.0 1.0 2.0 12421 1101.0 38 6 0.0 4.0 3.0 12422 5258.0 118 17 3.0 5.0 5.0 12423 2640.0 23 10 0.0 5.0 4.0 12424 3346.0 185 4 3.0 3.0 4.0 12425 1752.0 101 2 2.0 2.0 4.0 12426 1354.0 217 2 3.0 2.0 3.0 12427 1131.0 44 3 1.0 3.0 3.0 12428 7992.0 48 9 1.0 5.0 5.0 12429 7488.0 32 8 0.0 4.0 5.0 12430 256.0 66 1 2.0 1.0 1.0 12431 10905.0 58 27 1.0 5.0 5.0 12432 5069.0 65 5 2.0 4.0 5.0 12433 20577.0 23 10 0.0 5.0 5.0 12434 1113.0 107 3 2.0 3.0 3.0 12435 13931.0 102 6 2.0 4.0 5.0 12436 511.0 122 1 3.0 1.0 2.0 12437 12702.0 24 39 0.0 5.0 5.0 12438 2909.0 37 2 0.0 2.0 4.0 12439 1164.0 705 2 5.0 2.0 3.0 12440 864.0 485 2 5.0 2.0 2.0 12441 352.0 389 1 4.0 1.0 1.0 12442 172.0 26 1 0.0 1.0 0.0 12443 487.0 755 1 5.0 1.0 2.0 12444 5024.0 44 5 1.0 4.0 5.0 12445 437.0 45 2 1.0 2.0 1.0 12446 1014.0 80 1 2.0 1.0 3.0 12447 479.0 266 1 3.0 1.0 2.0 12448 451.0 67 1 2.0 1.0 1.0 12449 4841.0 45 6 1.0 4.0 5.0 12450 693.0 179 5 3.0 4.0 2.0 12451 12708.0 33 8 0.0 4.0 5.0 12452 432.0 39 2 1.0 2.0 1.0 12453 712.0 157 1 3.0 1.0 2.0 12454 16681.0 79 5 2.0 4.0 5.0 12455 5900.0 96 10 2.0 5.0 5.0 12456 3189.0 67 3 2.0 3.0 4.0 12457 3000.0 89 8 2.0 4.0 4.0 12458 1790.0 94 4 2.0 3.0 4.0 12459 453.0 628 1 5.0 1.0 1.0 12460 329.0 479 1 5.0 1.0 1.0 12461 832.0 117 2 2.0 2.0 2.0 12462 1197.0 41 2 1.0 2.0 3.0 12463 1350.0 76 4 2.0 3.0 3.0 12464 1215.0 33 8 0.0 4.0 3.0 12465 744.0 40 2 1.0 2.0 2.0 12466 57.0 705 1 5.0 1.0 0.0 12467 135.0 446 1 4.0 1.0 0.0 12468 3681.0 210 5 3.0 4.0 5.0 12469 3082.0 545 2 5.0 2.0 4.0 12470 215.0 712 1 5.0 1.0 0.0 12471 39995.0 25 79 0.0 5.0 5.0 12472 17947.0 55 19 1.0 5.0 5.0 12473 3955.0 52 5 1.0 4.0 5.0 12474 12479.0 40 31 1.0 5.0 5.0 12475 1157.0 76 2 2.0 2.0 3.0 124

#### Reverse the recency_score value; the lower the recency value, the higher the score should be (QuantileDiscretizer gives a high score to high values)

In [0]:
rfm_df = rfm_df.select("customer_id", "recency", "frequency", "monetary", abs(rfm_df.recency_score - 5).alias("recency_score"), "frequency_score", "monetary_score")
display(rfm_df.orderBy("customer_id"))

customer_id recency frequency monetary recency_score frequency_score monetary_score 12346 348 12 77551.0 1.0 5.0 5.0 12347 25 8 5641.0 5.0 4.0 5.0 12348 98 5 2016.0 3.0 4.0 4.0 12349 41 4 4454.0 4.0 3.0 5.0 12350 333 1 334.0 1.0 1.0 1.0 12351 398 1 304.0 1.0 1.0 1.0 12352 59 10 2855.0 4.0 5.0 4.0 12353 227 2 409.0 2.0 2.0 1.0 12354 255 1 1086.0 2.0 1.0 3.0 12355 237 2 950.0 2.0 2.0 3.0 12356 45 6 6371.0 4.0 4.0 5.0 12357 56 3 18287.0 4.0 3.0 5.0 12358 24 5 3893.0 5.0 4.0 5.0 12359 80 10 8990.0 3.0 5.0 5.0 12360 75 8 4267.0 3.0 4.0 5.0 12361 310 4 516.0 1.0 3.0 2.0 12362 26 11 5381.0 5.0 5.0 5.0 12363 132 2 549.0 2.0 2.0 2.0 12364 30 4 1315.0 5.0 3.0 3.0 12365 314 2 644.0 1.0 2.0 2.0 12366 656 1 500.0 0.0 1.0 2.0 12367 27 1 170.0 5.0 1.0 0.0 12368 651 1 920.0 0.0 1.0 3.0 12369 436 3 1800.0 1.0 3.0 4.0 12370 74 7 4339.0 3.0 4.0 5.0 12371 67 5 4083.0 3.0 4.0 5.0 12372 94 3 1302.0 3.0 3.0 3.0 12373 334 3 1137.0 1.0 3.0 3.0 12374 48 4 2992.0 4.0 3.0 4.0 12375 33 2 461.0 5.0 2.0 1.0 12376 412 2 504.0 1.0 2.0 2.0 12377 338 4 3428.0 1.0 3.0 4.0 12378 152 2 5440.0 2.0 2.0 5.0 12379 104 4 1631.0 3.0 3.0 3.0 12380 44 11 9727.0 4.0 5.0 5.0 12381 27 5 1850.0 5.0 4.0 4.0 12383 207 5 1856.0 2.0 4.0 4.0 12384 51 2 589.0 4.0 2.0 2.0 12385 402 1 1950.0 1.0 1.0 4.0 12386 360 2 661.0 1.0 2.0 2.0 12387 438 1 143.0 1.0 1.0 0.0 12388 38 8 3903.0 5.0 4.0 5.0 12389 425 3 1436.0 1.0 3.0 3.0 12390 102 1 554.0 3.0 1.0 2.0 12391 44 6 2799.0 4.0 4.0 4.0 12392 614 1 236.0 0.0 1.0 0.0 12393 95 6 2413.0 3.0 4.0 4.0 12394 86 2 1274.0 3.0 2.0 3.0 12395 42 15 5080.0 4.0 5.0 5.0 12396 697 1 937.0 0.0 1.0 3.0 12397 58 2 2413.0 4.0 2.0 4.0 12398 68 1 1639.0 3.0 1.0 3.0 12399 142 7 1959.0 2.0 4.0 4.0 12400 437 1 205.0 1.0 1.0 0.0 12401 326 1 85.0 1.0 1.0 0.0 12402 346 6 1271.0 1.0 4.0 3.0 12403 72 2 1020.0 3.0 2.0 3.0 12404 705 1 63.0 0.0 1.0 0.0 12405 171 1 1714.0 2.0 1.0 4.0 12406 45 5 6394.0 4.0 4.0 5.0 12407 72 5 1712.0 3.0 4.0 4.0 12408 55 8 4356.0 4.0 4.0 5.0 12409 101 7 23426.0 3.0 4.0 5.0 12410 331 3 1016.0 1.0 3.0 3.0 12411 432 1 354.0 1.0 1.0 1.0 12412 97 6 2595.0 3.0 4.0 4.0 12413 89 4 1008.0 3.0 3.0 3.0 12414 240 6 954.0 2.0 4.0 3.0 12415 47 28 144470.0 4.0 5.0 5.0 12416 679 1 202.0 0.0 1.0 0.0 12417 26 20 6857.0 5.0 5.0 5.0 12418 135 4 2027.0 2.0 3.0 4.0 12419 408 2 505.0 1.0 2.0 2.0 12420 86 1 604.0 3.0 1.0 2.0 12421 38 6 1101.0 5.0 4.0 3.0 12422 118 17 5258.0 2.0 5.0 5.0 12423 23 10 2640.0 5.0 5.0 4.0 12424 185 4 3346.0 2.0 3.0 4.0 12425 101 2 1752.0 3.0 2.0 4.0 12426 217 2 1354.0 2.0 2.0 3.0 12427 44 3 1131.0 4.0 3.0 3.0 12428 48 9 7992.0 4.0 5.0 5.0 12429 32 8 7488.0 5.0 4.0 5.0 12430 66 1 256.0 3.0 1.0 1.0 12431 58 27 10905.0 4.0 5.0 5.0 12432 65 5 5069.0 3.0 4.0 5.0 12433 23 10 20577.0 5.0 5.0 5.0 12434 107 3 1113.0 3.0 3.0 3.0 12435 102 6 13931.0 3.0 4.0 5.0 12436 122 1 511.0 2.0 1.0 2.0 12437 24 39 12702.0 5.0 5.0 5.0 12438 37 2 2909.0 5.0 2.0 4.0 12439 705 2 1164.0 0.0 2.0 3.0 12440 485 2 864.0 0.0 2.0 2.0 12441 389 1 352.0 1.0 1.0 1.0 12442 26 1 172.0 5.0 1.0 0.0 12443 755 1 487.0 0.0 1.0 2.0 12444 44 5 5024.0 4.0 4.0 5.0 12445 45 2 437.0 4.0 2.0 1.0 12446 80 1 1014.0 3.0 1.0 3.0 12447 266 1 479.0 2.0 1.0 2.0 12448 67 1 451.0 3.0 1.0 1.0 12449 45 6 4841.0 4.0 4.0 5.0 12450 179 5 693.0 2.0 4.0 2.0 12451 33 8 12708.0 5.0 4.0 5.0 12452 39 2 432.0 4.0 2.0 1.0 12453 157 1 712.0 2.0 1.0 2.0 12454 79 5 16681.0 3.0 4.0 5.0 12455 96 10 5900.0 3.0 5.0 5.0 12456 67 3 3189.0 3.0 3.0 4.0 12457 89 8 3000.0 3.0 4.0 4.0 12458 94 4 1790.0 3.0 3.0 4.0 12459 628 1 453.0 0.0 1.0 1.0 12460 479 1 329.0 0.0 1.0 1.0 12461 117 2 832.0 3.0 2.0 2.0 12462 41 2 1197.0 4.0 2.0 3.0 12463 76 4 1350.0 3.0 3.0 3.0 12464 33 8 1215.0 5.0 4.0 3.0 12465 40 2 744.0 4.0 2.0 2.0 12466 705 1 57.0 0.0 1.0 0.0 12467 446 1 135.0 1.0 1.0 0.0 12468 210 5 3681.0 2.0 4.0 5.0 12469 545 2 3082.0 0.0 2.0 4.0 12470 712 1 215.0 0.0 1.0 0.0 12471 25 79 39995.0 5.0 5.0 5.0 12472 55 19 17947.0 4.0 5.0 5.0 12473 52 5 3955.0 4.0 4.0 5.0 12474 40 31 12479.0 4.0 5.0 5.0 12475 76 2 1157.0 3.0 2.0 3.0 124

#### Unify RFM scores

In [0]:
rfm_df = rfm_df.withColumn("rfm_score", concat(col("recency_score").cast("int"), col("frequency_score").cast("int"), col("monetary_score").cast("int")))
display(rfm_df.orderBy("customer_id"))

customer_id recency frequency monetary recency_score frequency_score monetary_score rfm_score 12346 348 12 77551.0 1.0 5.0 5.0 155 12347 25 8 5641.0 5.0 4.0 5.0 545 12348 98 5 2016.0 3.0 4.0 4.0 344 12349 41 4 4454.0 4.0 3.0 5.0 435 12350 333 1 334.0 1.0 1.0 1.0 111 12351 398 1 304.0 1.0 1.0 1.0 111 12352 59 10 2855.0 4.0 5.0 4.0 454 12353 227 2 409.0 2.0 2.0 1.0 221 12354 255 1 1086.0 2.0 1.0 3.0 213 12355 237 2 950.0 2.0 2.0 3.0 223 12356 45 6 6371.0 4.0 4.0 5.0 445 12357 56 3 18287.0 4.0 3.0 5.0 435 12358 24 5 3893.0 5.0 4.0 5.0 545 12359 80 10 8990.0 3.0 5.0 5.0 355 12360 75 8 4267.0 3.0 4.0 5.0 345 12361 310 4 516.0 1.0 3.0 2.0 132 12362 26 11 5381.0 5.0 5.0 5.0 555 12363 132 2 549.0 2.0 2.0 2.0 222 12364 30 4 1315.0 5.0 3.0 3.0 533 12365 314 2 644.0 1.0 2.0 2.0 122 12366 656 1 500.0 0.0 1.0 2.0 012 12367 27 1 170.0 5.0 1.0 0.0 510 12368 651 1 920.0 0.0 1.0 3.0 013 12369 436 3 1800.0 1.0 3.0 4.0 134 12370 74 7 4339.0 3.0 4.0 5.0 345 12371 67 5 4083.0 3.0 4.0 5.0 345 12372 94 3 1302.0 3.0 3.0 3.0 333 12373 334 3 1137.0 1.0 3.0 3.0 133 12374 48 4 2992.0 4.0 3.0 4.0 434 12375 33 2 461.0 5.0 2.0 1.0 521 12376 412 2 504.0 1.0 2.0 2.0 122 12377 338 4 3428.0 1.0 3.0 4.0 134 12378 152 2 5440.0 2.0 2.0 5.0 225 12379 104 4 1631.0 3.0 3.0 3.0 333 12380 44 11 9727.0 4.0 5.0 5.0 455 12381 27 5 1850.0 5.0 4.0 4.0 544 12383 207 5 1856.0 2.0 4.0 4.0 244 12384 51 2 589.0 4.0 2.0 2.0 422 12385 402 1 1950.0 1.0 1.0 4.0 114 12386 360 2 661.0 1.0 2.0 2.0 122 12387 438 1 143.0 1.0 1.0 0.0 110 12388 38 8 3903.0 5.0 4.0 5.0 545 12389 425 3 1436.0 1.0 3.0 3.0 133 12390 102 1 554.0 3.0 1.0 2.0 312 12391 44 6 2799.0 4.0 4.0 4.0 444 12392 614 1 236.0 0.0 1.0 0.0 010 12393 95 6 2413.0 3.0 4.0 4.0 344 12394 86 2 1274.0 3.0 2.0 3.0 323 12395 42 15 5080.0 4.0 5.0 5.0 455 12396 697 1 937.0 0.0 1.0 3.0 013 12397 58 2 2413.0 4.0 2.0 4.0 424 12398 68 1 1639.0 3.0 1.0 3.0 313 12399 142 7 1959.0 2.0 4.0 4.0 244 12400 437 1 205.0 1.0 1.0 0.0 110 12401 326 1 85.0 1.0 1.0 0.0 110 12402 346 6 1271.0 1.0 4.0 3.0 143 12403 72 2 1020.0 3.0 2.0 3.0 323 12404 705 1 63.0 0.0 1.0 0.0 010 12405 171 1 1714.0 2.0 1.0 4.0 214 12406 45 5 6394.0 4.0 4.0 5.0 445 12407 72 5 1712.0 3.0 4.0 4.0 344 12408 55 8 4356.0 4.0 4.0 5.0 445 12409 101 7 23426.0 3.0 4.0 5.0 345 12410 331 3 1016.0 1.0 3.0 3.0 133 12411 432 1 354.0 1.0 1.0 1.0 111 12412 97 6 2595.0 3.0 4.0 4.0 344 12413 89 4 1008.0 3.0 3.0 3.0 333 12414 240 6 954.0 2.0 4.0 3.0 243 12415 47 28 144470.0 4.0 5.0 5.0 455 12416 679 1 202.0 0.0 1.0 0.0 010 12417 26 20 6857.0 5.0 5.0 5.0 555 12418 135 4 2027.0 2.0 3.0 4.0 234 12419 408 2 505.0 1.0 2.0 2.0 122 12420 86 1 604.0 3.0 1.0 2.0 312 12421 38 6 1101.0 5.0 4.0 3.0 543 12422 118 17 5258.0 2.0 5.0 5.0 255 12423 23 10 2640.0 5.0 5.0 4.0 554 12424 185 4 3346.0 2.0 3.0 4.0 234 12425 101 2 1752.0 3.0 2.0 4.0 324 12426 217 2 1354.0 2.0 2.0 3.0 223 12427 44 3 1131.0 4.0 3.0 3.0 433 12428 48 9 7992.0 4.0 5.0 5.0 455 12429 32 8 7488.0 5.0 4.0 5.0 545 12430 66 1 256.0 3.0 1.0 1.0 311 12431 58 27 10905.0 4.0 5.0 5.0 455 12432 65 5 5069.0 3.0 4.0 5.0 345 12433 23 10 20577.0 5.0 5.0 5.0 555 12434 107 3 1113.0 3.0 3.0 3.0 333 12435 102 6 13931.0 3.0 4.0 5.0 345 12436 122 1 511.0 2.0 1.0 2.0 212 12437 24 39 12702.0 5.0 5.0 5.0 555 12438 37 2 2909.0 5.0 2.0 4.0 524 12439 705 2 1164.0 0.0 2.0 3.0 023 12440 485 2 864.0 0.0 2.0 2.0 022 12441 389 1 352.0 1.0 1.0 1.0 111 12442 26 1 172.0 5.0 1.0 0.0 510 12443 755 1 487.0 0.0 1.0 2.0 012 12444 44 5 5024.0 4.0 4.0 5.0 445 12445 45 2 437.0 4.0 2.0 1.0 421 12446 80 1 1014.0 3.0 1.0 3.0 313 12447 266 1 479.0 2.0 1.0 2.0 212 12448 67 1 451.0 3.0 1.0 1.0 311 12449 45 6 4841.0 4.0 4.0 5.0 445 12450 179 5 693.0 2.0 4.0 2.0 242 12451 33 8 12708.0 5.0 4.0 5.0 545 12452 39 2 432.0 4.0 2.0 1.0 421 12453 157 1 712.0 2.0 1.0 2.0 212 12454 79 5 16681.0 3.0 4.0 5.0 345 12455 96 10 5900.0 3.0 5.0 5.0 355 12456 67 3 3189.0 3.0 3.0 4.0 334 12457 89 8 3000.0 3.0 4.0 4.0 344 12458 94 4 1790.0 3.0 3.0 4.0 334 12459 628 1 453.0 0.0 1.0 1.0 011 12460 479 1 329.0 0.0 1.0 1

#### Create segments based on RFM scores; use regular expressions to replace RFM scores with corresponding segments

In [0]:
rfm_df = rfm_df.withColumn("segment_tmp", concat(col("recency_score").cast("int"), col("frequency_score").cast("int")))

In [0]:
rfm_seg_df = rfm_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment_tmp"), r'[0-2][0-2]', "Hibernating").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'[0-2][3-4]', "At Risk").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'[0-2]5', "Can\'t Lose").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'3[0-2]', "About to Sleep").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'33', "Need Attention").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'[3-4][4-5]', "Loyal Customers").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'4[0-1]', "Promising").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'5[0-1]', "New Customers").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'[4-5][2-3]', "Potential Loyalists").alias("segment"))
rfm_seg_df = rfm_seg_df.select("customer_id", "recency", "frequency", "monetary", "recency_score", "frequency_score", "monetary_score", "rfm_score", 
                                                      regexp_replace(col("segment"), r'5[4-5]', "Champions").alias("segment"))
display(rfm_seg_df.orderBy("customer_id"))

customer_id recency frequency monetary recency_score frequency_score monetary_score rfm_score segment 12346 348 12 77551.0 1.0 5.0 5.0 155 Can't Lose 12347 25 8 5641.0 5.0 4.0 5.0 545 Champions 12348 98 5 2016.0 3.0 4.0 4.0 344 Loyal Customers 12349 41 4 4454.0 4.0 3.0 5.0 435 Potential Loyalists 12350 333 1 334.0 1.0 1.0 1.0 111 Hibernating 12351 398 1 304.0 1.0 1.0 1.0 111 Hibernating 12352 59 10 2855.0 4.0 5.0 4.0 454 Loyal Customers 12353 227 2 409.0 2.0 2.0 1.0 221 Hibernating 12354 255 1 1086.0 2.0 1.0 3.0 213 Hibernating 12355 237 2 950.0 2.0 2.0 3.0 223 Hibernating 12356 45 6 6371.0 4.0 4.0 5.0 445 Loyal Customers 12357 56 3 18287.0 4.0 3.0 5.0 435 Potential Loyalists 12358 24 5 3893.0 5.0 4.0 5.0 545 Champions 12359 80 10 8990.0 3.0 5.0 5.0 355 Loyal Customers 12360 75 8 4267.0 3.0 4.0 5.0 345 Loyal Customers 12361 310 4 516.0 1.0 3.0 2.0 132 At Risk 12362 26 11 5381.0 5.0 5.0 5.0 555 Champions 12363 132 2 549.0 2.0 2.0 2.0 222 Hibernating 12364 30 4 1315.0 5.0 3.0 3.0 533 Potential Loyalists 12365 314 2 644.0 1.0 2.0 2.0 122 Hibernating 12366 656 1 500.0 0.0 1.0 2.0 012 Hibernating 12367 27 1 170.0 5.0 1.0 0.0 510 New Customers 12368 651 1 920.0 0.0 1.0 3.0 013 Hibernating 12369 436 3 1800.0 1.0 3.0 4.0 134 At Risk 12370 74 7 4339.0 3.0 4.0 5.0 345 Loyal Customers 12371 67 5 4083.0 3.0 4.0 5.0 345 Loyal Customers 12372 94 3 1302.0 3.0 3.0 3.0 333 Need Attention 12373 334 3 1137.0 1.0 3.0 3.0 133 At Risk 12374 48 4 2992.0 4.0 3.0 4.0 434 Potential Loyalists 12375 33 2 461.0 5.0 2.0 1.0 521 Potential Loyalists 12376 412 2 504.0 1.0 2.0 2.0 122 Hibernating 12377 338 4 3428.0 1.0 3.0 4.0 134 At Risk 12378 152 2 5440.0 2.0 2.0 5.0 225 Hibernating 12379 104 4 1631.0 3.0 3.0 3.0 333 Need Attention 12380 44 11 9727.0 4.0 5.0 5.0 455 Loyal Customers 12381 27 5 1850.0 5.0 4.0 4.0 544 Champions 12383 207 5 1856.0 2.0 4.0 4.0 244 At Risk 12384 51 2 589.0 4.0 2.0 2.0 422 Potential Loyalists 12385 402 1 1950.0 1.0 1.0 4.0 114 Hibernating 12386 360 2 661.0 1.0 2.0 2.0 122 Hibernating 12387 438 1 143.0 1.0 1.0 0.0 110 Hibernating 12388 38 8 3903.0 5.0 4.0 5.0 545 Champions 12389 425 3 1436.0 1.0 3.0 3.0 133 At Risk 12390 102 1 554.0 3.0 1.0 2.0 312 About to Sleep 12391 44 6 2799.0 4.0 4.0 4.0 444 Loyal Customers 12392 614 1 236.0 0.0 1.0 0.0 010 Hibernating 12393 95 6 2413.0 3.0 4.0 4.0 344 Loyal Customers 12394 86 2 1274.0 3.0 2.0 3.0 323 About to Sleep 12395 42 15 5080.0 4.0 5.0 5.0 455 Loyal Customers 12396 697 1 937.0 0.0 1.0 3.0 013 Hibernating 12397 58 2 2413.0 4.0 2.0 4.0 424 Potential Loyalists 12398 68 1 1639.0 3.0 1.0 3.0 313 About to Sleep 12399 142 7 1959.0 2.0 4.0 4.0 244 At Risk 12400 437 1 205.0 1.0 1.0 0.0 110 Hibernating 12401 326 1 85.0 1.0 1.0 0.0 110 Hibernating 12402 346 6 1271.0 1.0 4.0 3.0 143 At Risk 12403 72 2 1020.0 3.0 2.0 3.0 323 About to Sleep 12404 705 1 63.0 0.0 1.0 0.0 010 Hibernating 12405 171 1 1714.0 2.0 1.0 4.0 214 Hibernating 12406 45 5 6394.0 4.0 4.0 5.0 445 Loyal Customers 12407 72 5 1712.0 3.0 4.0 4.0 344 Loyal Customers 12408 55 8 4356.0 4.0 4.0 5.0 445 Loyal Customers 12409 101 7 23426.0 3.0 4.0 5.0 345 Loyal Customers 12410 331 3 1016.0 1.0 3.0 3.0 133 At Risk 12411 432 1 354.0 1.0 1.0 1.0 111 Hibernating 12412 97 6 2595.0 3.0 4.0 4.0 344 Loyal Customers 12413 89 4 1008.0 3.0 3.0 3.0 333 Need Attention 12414 240 6 954.0 2.0 4.0 3.0 243 At Risk 12415 47 28 144470.0 4.0 5.0 5.0 455 Loyal Customers 12416 679 1 202.0 0.0 1.0 0.0 010 Hibernating 12417 26 20 6857.0 5.0 5.0 5.0 555 Champions 12418 135 4 2027.0 2.0 3.0 4.0 234 At Risk 12419 408 2 505.0 1.0 2.0 2.0 122 Hibernating 12420 86 1 604.0 3.0 1.0 2.0 312 About to Sleep 12421 38 6 1101.0 5.0 4.0 3.0 543 Champions 12422 118 17 5258.0 2.0 5.0 5.0 255 Can't Lose 12423 23 10 2640.0 5.0 5.0 4.0 554 Champions 12424 185 4 3346.0 2.0 3.0 4.0 234 At Risk 12425 101 2 1752.0 3.0 2.0 4.0 324 About to Sleep 12426 217 2 1354.0 2.0 2.0 3.0 223 Hibernating 12427 44 3 1131.0 4.0 3.0 3.0 433 Potential Loyalists 12428 48 9 7992.0 4.0 5.0 5.0 455 Loyal Customer

#### Get the RFM mean and count for each segment

In [0]:
seg_df = rfm_seg_df.groupBy("segment").agg(mean("recency").cast("int").alias("recency_mean"), count("recency").alias("recency_count"),
                                            mean("frequency").cast("int").alias("frequency_mean"), count("frequency").alias("frequency_count"),
                                            mean("monetary").cast("int").alias("monetary_mean"), count("monetary").alias("monetary_count")) 
display(seg_df.orderBy("segment"))

segment recency_mean recency_count frequency_mean frequency_count monetary_mean monetary_count About to Sleep 87 334 1 334 544 334 At Risk 314 948 4 948 1613 948 Can't Lose 239 148 15 148 7077 148 Champions 29 689 21 689 12299 689 Hibernating 442 1853 1 1853 457 1853 Loyal Customers 65 955 11 955 4877 955 Need Attention 85 228 3 228 1264 228 New Customers 31 56 1 56 383 56 Potential Loyalists 41 557 2 557 1298 557 Promising 50 110 1 110 312 110

In [0]:
seg_complex_df = seg_df.select("segment", struct("recency_mean", "recency_count").alias("recency"),
                                          struct("frequency_mean", "frequency_count").alias("frequency"),
                                          struct("monetary_mean", "monetary_count").alias("monetary"))
                     
display(seg_complex_df.orderBy("segment"))

segment recency frequency monetary About to Sleep List(87, 334) List(1, 334) List(544, 334) At Risk List(314, 948) List(4, 948) List(1613, 948) Can't Lose List(239, 148) List(15, 148) List(7077, 148) Champions List(29, 689) List(21, 689) List(12299, 689) Hibernating List(442, 1853) List(1, 1853) List(457, 1853) Loyal Customers List(65, 955) List(11, 955) List(4877, 955) Need Attention List(85, 228) List(3, 228) List(1264, 228) New Customers List(31, 56) List(1, 56) List(383, 56) Potential Loyalists List(41, 557) List(2, 557) List(1298, 557) Promising List(50, 110) List(1, 110) List(312, 110)